---
title: "iPSC Motor Neurons and Spinal Cord Cell Refs Data"
author: "Jenny L Smith"
date: "`R Sys.Date()`"
output: html_document
---

# Set-up 

```{r set-up, echo=FALSE}
knitr::opts_knit$set(root.dir = rprojroot::find_rstudio_root_file())
knitr::opts_chunk$set(tidy.opts=list(width.cutoff=80),
                      tidy=TRUE,
                      fig.align='center',
                      fig.width = 10, fig.height = 10,
                      eval = TRUE)

options(stringsAsFactors = FALSE, max.print = 100)
table = function (..., useNA = 'ifany') base::table(..., useNA = useNA)
```

```{r message=FALSE,warning=FALSE}
library(dplyr)
library(tidyr)
library(magrittr)
library(stringr)
library(tibble)
library(glue)

library(ggplot2)
library(plotly)
library(RColorBrewer)

library(Seurat)
library(SeuratDisk)
library(SeuratObject)
library(SeuratWrappers)
library(Matrix)
library(SeuratObject)
library(SeuratData)
library(SeuratWrappers)

library(scDblFinder)
library(SoupX)
```

```{r}
conflicted::conflict_prefer("select", "dplyr")
conflicted::conflict_prefer("filter","dplyr")
conflicted::conflict_prefer("saveRDS","base")
```


# Parallelization 

```{r}
library(future)
# check the current active plan
plan()

plan("multisession", workers = 16)
plan()
options(future.globals.maxSize = 50000 * 1024^2)
```

# Reference Data Set(s)

1. https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE136719
2. https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE138121
3. https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE171892
3-1. https://juliendelile.github.io/Antler/
3-2. https://shiny.crick.ac.uk/scviewer/neuraltube/

4. Yadav 2023 
  https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE190442
  https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE222322

```{bash eval=FALSE}
wget https://ftp.ncbi.nlm.nih.gov/geo/series/GSE190nnn/GSE190442/suppl/GSE190442_aggregated_counts_postqc.csv.gz
wget https://ftp.ncbi.nlm.nih.gov/geo/series/GSE190nnn/GSE190442/suppl/GSE190442_aggregated_metadata_postqc.csv.gz

rclone copy --http-url https://zenodo.org :http:record/4546932 . -P
rclone copy --http-url http://celltypes.brain-map.org :http:api/v2/well_known_file_download/694416044 . -P
rclone copy --http-url http://celltypes.brain-map.org :http:api/v2/well_known_file_download/502175284 . -P
rclone copy refseq:/geo/series/GSE190nnn/GSE190442/suppl/GSE190442_RAW.tar . -P
```

```{bash eval=FALSE}
cd resources/cell_refs/allen_institute/M1

# https://azimuth.hubmapconsortium.org/references/#Human%20-%20Motor%20Cortex
#wget https://seurat.nygenome.org/azimuth/demo_datasets/allen_m1c_2019_ssv4.rds

rclone copy --http-url https://brainmapportal-live-4cc80a57cd6e400d854-f7fdcae.divio-media.net :http:filer_public/70/32/70326830-e306-4743-a02c-a8da5bf9eb56/readme-m1-10.txt  . -P

rclone copy --http-url https://brainmapportal-live-4cc80a57cd6e400d854-f7fdcae.divio-media.net \
  :http:filer_public/0c/0c/0c0c882d-1c31-40a9-8039-3bf2706a77cd/sample-exp_component_mapping_human_10x_apr2020.zip . -P 

rclone copy --http-url https://idk-etl-prod-download-bucket.s3.amazonaws.com :http:aibs_human_m1_10x/metadata.csv . -P
rclone copy --http-url https://idk-etl-prod-download-bucket.s3.amazonaws.com :http:aibs_human_m1_10x/matrix.csv . -P
```


# Genome Refs

```{r eval=FALSE}
edb <- ensembldb::EnsDb("resources/genome/Homo_sapiens.GRCh38.98.sqlite")

# edb

gids <- ensembldb::keys(edb, keytype = "GENEID")

idmap <- AnnotationDbi::select(edb, keytype="GENEID", 
                      keys = gids, 
                      columns = c("GENEID","GENEBIOTYPE","GENENAME","SYMBOL")) %>% 
  janitor::clean_names()

head(idmap)
```

```{r eval=FALSE}
MTG_genes <- read.csv("resources/cell_refs/allen_institute/human_MTG_2018-06-14_genes-rows.csv")

head(MTG_genes)
```

# Preprocessing

## Allen Institute:

* This data set includes single-nucleus transcriptomes from 76,533 total nuclei derived from 2 post-mortem human brain specimens, to survey 
cell type diversity in the primary motor cortex (M1C or M1).
*  default 10x Cell Ranger v3

* https://brainmapportal-live-4cc80a57cd6e400d854-f7fdcae.divio-media.net/filer_public/30/80/30803016-9515-45fd-a6af-8595902c176f/readme_human_ctx.txt 

* https://github.com/satijalab/azimuth-references/tree/master/human_motorcortex 
* https://knowledge.brain-map.org/data/8GU8FPSNLBUBCJWQDZ5/summary

```{r}
mc_meta <- read.csv("resources/cell_refs/allen_institute/M1/metadata.csv")
rownames(mc_meta) <- mc_meta$sample_name

head(mc_meta)
```

```{r}
az_mc_meta <- read.csv("resources/cell_refs/allen_institute/allen_institute_azimuth_subclass_marker_gene_list.csv") %>% 
  janitor::clean_names()

head(az_mc_meta)
```

```{r}
mc_meta %>% 
  group_by(cluster_label) %>% 
  dplyr::count() %>% 
  arrange(desc(n))
```

```{r eval=FALSE}
ref_mc_counts <- data.table::fread("resources/cell_refs/allen_institute/M1/matrix.csv") 


ref_mc_counts <- as.data.frame(ref_mc_counts)
rownames(ref_mc_counts) <- ref_mc_counts$sample_name
ref_mc_counts <- t(as.matrix(ref_mc_counts[,-1]))

dim(ref_mc_counts) #76533 50282
head(ref_mc_counts[,1:5])
```

```{r eval=FALSE}
ref_motor_cortex <- CreateSeuratObject(counts = ref_mc_counts,
                                   meta.data = mc_meta,
                                   assay="RNA",
                                   project="Allen_M1") 
rm(ref_mc_counts)
rm(mc_meta)
gc()

ref_motor_cortex <- ref_motor_cortex %>% 
  PercentageFeatureSet(
                         assay="RNA",
                         pattern = "^MT-",
                         col.name = "percent.mt") %>% 
  PercentageFeatureSet(
                       assay="RNA",
                       pattern = "^RP[SL]", 
                       col.name = "percent.ribo") %>% 
  SCTransform(
              vst.flavor = "v2", 
              method = "glmGamPoi",
              verbose = T) %>%
    RunPCA(npcs = 50, verbose = T) %>%
    RunUMAP(reduction = "pca", 
            dims = 1:50,
            n.components = 3, 
            min.dist = 0.3,
            verbose = T) %>%
    FindNeighbors(reduction = "pca",
                  dims = 1:50, verbose = T) %>%
    FindClusters(resolution = 0.8, verbose = T)

saveRDS(ref_motor_cortex,"resources/cell_refs/allen_institute/M1/human_M1_normalized_seurat_object.RDS")
```

```{r}
Idents(ref_motor_cortex) <- ref_motor_cortex$class
```

### MTG 

Not using the MTG dataset as M1 is larger and more comprehensice

```{r}
ref_mc_meta <- read.csv("resources/cell_refs/allen_institute/human_MTG_2018-06-14_samples-columns.csv")

head(ref_mc_meta)
tail(ref_mc_meta)
# dim(ref_mc_meta) # 15,928    34

# Assay data with 50281 features for 6235 cells
# First 10 features:
#  3.8-1.2, 3.8-1.3, 3.8-1.4, 3.8-1.5, 5-HT3C2, A1BG, A1BG-AS1, A1CF, A2M, A2M-AS1 
ref_counts="resources/cell_refs/allen_institute/allen_m1c_2019_ssv4.rds"
ref_motor_cortex <- readRDS(ref_counts)

dim(ref_motor_cortex) #42920  6,235
# head(ref_motor_cortex@meta.data)
dim(ref_motor_cortex@meta.data) #6235   15
# head(rownames(ref_motor_cortex), n=20)

ref_motor_cortex@meta.data %>% 
  group_by(cluster) %>% 
  dplyr::count()

ref_mc_meta %>% 
  group_by(cluster) %>% 
  dplyr::count()

table(ref_motor_cortex@meta.data$class)
table(ref_motor_cortex@meta.data$subclass)

grep("^MT-", rownames(ref_motor_cortex)) #MT genes are removed?
grep("^RP[SL]", rownames(ref_motor_cortex), value=TRUE)
```


## Yadav:

fastq files were run through 10x Genomics Cellranger v3.0.2
output bam files from cellranger were processed using BAMboozle for anonymization

* Genome_build: hg38
Supplementary_files_format_and_content: Counts tables for all barcodes in standard hdf5 output from Cellranger


* https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSM5723843

```{r}
sc_pts <- read.csv("resources/cell_refs/Yadav_2023/GSE222322_donor_to_patient_ids.csv")


sc_meta <- read.csv("resources/cell_refs/Yadav_2023/GSE190442_aggregated_metadata_postqc.csv") %>% 
  mutate(barcode=X) %>% 
  left_join(., sc_pts, by=c("sample"="Patient_ID")) %>% 
  column_to_rownames("X") 
  

head(sc_meta)
# table(sc_meta$subtype_annotation)
dim(sc_meta) # 55289     7
```

```{r}
sc_meta %>% 
  group_by(subtype_annotation) %>% 
  dplyr::count() %>% 
  arrange(desc(n))

# Motoneurons	1331 only
```

```{r eval=FALSE}
ref_counts_sc <- data.table::fread("resources/cell_refs/Yadav_2023/GSE190442_aggregated_counts_postqc.csv")
ref_counts_sc <- as.data.frame(ref_counts_sc)
rownames(ref_counts_sc) <- ref_counts_sc$V1
ref_mat_sc <- as.matrix(ref_counts_sc[,-1])
head(ref_mat_sc[,1:5])
dim(ref_mat_sc) #22238 55,290

ref_cell_meta=read.csv("resources/cell_refs/Yadav_2023/GSE190442_aggregated_metadata_postqc.csv", 
                       row.names = 1)
dim(ref_cell_meta) # 55289     6

ref_spinal_cord <- CreateSeuratObject(counts = ref_mat_sc,
                                   meta.data = ref_cell_meta,
                                   assay="RNA",
                                   project="yadav_2023")

ref_spinal_cord

# grep("^MT-", rownames(ref_spinal_cord)) #MT genes are removed?
# grep("^RP[SL]", rownames(ref_spinal_cord), value=TRUE)

# saveRDS(ref_spinal_cord, "resources/cell_refs/Yadav_2023/GSE190442_aggregated_postqc_seurat_obj.RDS")
```

```{r}
ref_spinal_cord <- readRDS("resources/cell_refs/Yadav_2023/GSE190442_aggregated_postqc_seurat_obj.RDS") %>% 
  PercentageFeatureSet(
                       assay="RNA",
                       pattern = "^MT-",
                       col.name = "percent.mt") %>% 
  PercentageFeatureSet(
                       assay="RNA",
                       pattern = "^RP[SL]", 
                       col.name = "percent.ribo") %>% 
  SCTransform( 
              vst.flavor = "v2", 
              method = "glmGamPoi",
              verbose = T) %>%
    RunPCA(npcs = 50, verbose = T) %>%
    RunUMAP(reduction = "pca", 
            dims = 1:50,
            n.components = 3, 
            min.dist = 0.3,
            verbose = T) %>%
    FindNeighbors(reduction = "pca",
                  dims = 1:50, verbose = T) %>%
    FindClusters(resolution = 0.8, verbose = T)


ref_spinal_cord <- AddMetaData(ref_spinal_cord, select(sc_meta, barcode:Visium_ID))
table(ref_spinal_cord@meta.data$top_level_annotation)
# saveRDS(ref_spinal_cord, "resources/cell_refs/Yadav_2023/GSE190442_aggregated_postqc_normalized_seurat_obj.RDS")
```


### Visualization and QC

```{r}

```


# Compare Gene Refs

```{r}
head(ref_motor_cortex@assays$RNA)
head(ref_spinal_cord@assays$RNA)

genes_common <- base::intersect(rownames(ref_motor_cortex), rownames(ref_spinal_cord))
length(genes_common) #20243 only in common :/
```

```{r}
missing_genes <- data.frame(gene_name=rownames(ref_motor_cortex),
           data_set="mc") %>% 
  full_join(., data.frame(gene_name=rownames(ref_spinal_cord),
                          data_set="sc"),
            by="gene_name")

missing_genes %>% 
  filter(is.na(data_set.x))
```

# Integrate

# Session Info

```{r}
session_info()
```

