---
title: "iPSC Motor Neurons and Spinal Cord Cell Refs Data"
author: "Jenny L Smith"
date: "`R Sys.Date()`"
output: html_document
---

# Set-up 

```{r set-up, echo=FALSE}
knitr::opts_knit$set(root.dir = rprojroot::find_rstudio_root_file())
knitr::opts_chunk$set(tidy.opts=list(width.cutoff=80),
                      tidy=TRUE,
                      fig.align='center',
                      fig.width = 10, fig.height = 10,
                      eval = TRUE)

options(stringsAsFactors = FALSE, max.print = 100)
table = function (..., useNA = 'ifany') base::table(..., useNA = useNA)
```

```{r message=FALSE,warning=FALSE}
library(dplyr)
library(tidyr)
library(magrittr)
library(stringr)
library(tibble)
library(glue)

library(ggplot2)
library(plotly)
library(RColorBrewer)

library(Seurat)
library(SeuratDisk)
library(SeuratObject)
library(SeuratWrappers)
library(Matrix)
library(SeuratObject)
library(SeuratData)
library(SeuratWrappers)

library(scDblFinder)
library(SoupX)
```

```{r}
conflicted::conflict_prefer("select", "dplyr")
conflicted::conflict_prefer("filter","dplyr")
conflicted::conflict_prefer("saveRDS","base")
```


# Parallelization 

```{r}
library(future)
# check the current active plan
plan()
```

```{r}
plan("multisession", workers = 16)
plan()
options(future.globals.maxSize = 60000 * 1024^2)
```


# Genome Refs

```{r}
edb <- ensembldb::EnsDb("resources/genome/Homo_sapiens.GRCh38.98.sqlite")

# edb
```

```{r}
gids <- ensembldb::keys(edb, keytype = "GENEID")

idmap <- AnnotationDbi::select(edb, keytype="GENEID", 
                      keys = gids, 
                      columns = c("GENEID","GENEBIOTYPE","GENENAME","SYMBOL")) %>% 
  janitor::clean_names()

head(idmap)
```

# Reference Data Set(s)

1. https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE136719
2. https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE138121
3. https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE171892
3-1. https://juliendelile.github.io/Antler/
3-2. https://shiny.crick.ac.uk/scviewer/neuraltube/

4. Yadav 2023 
  https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE190442
  https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE222322

```{bash eval=FALSE}
wget https://ftp.ncbi.nlm.nih.gov/geo/series/GSE190nnn/GSE190442/suppl/GSE190442_aggregated_counts_postqc.csv.gz
wget https://ftp.ncbi.nlm.nih.gov/geo/series/GSE190nnn/GSE190442/suppl/GSE190442_aggregated_metadata_postqc.csv.gz

rsync -av https://zenodo.org/record/4546932
wget https://seurat.nygenome.org/azimuth/demo_datasets/allen_m1c_2019_ssv4.rds
rsync -av http://celltypes.brain-map.org/api/v2/well_known_file_download/694416044
rsync -av http://celltypes.brain-map.org/api/v2/well_known_file_download/502175284
```

# Preprocessing

Allen Institute:
* The sequencing results were aligned to exons and introns in the GRCh38.p2 reference genome using the STAR algorithm, and aggregated intron and exon counts at the gene level were calculated.

* https://brainmapportal-live-4cc80a57cd6e400d854-f7fdcae.divio-media.net/filer_public/30/80/30803016-9515-45fd-a6af-8595902c176f/readme_human_ctx.txt 

```{r}
# Assay data with 50281 features for 6235 cells
# First 10 features:
#  3.8-1.2, 3.8-1.3, 3.8-1.4, 3.8-1.5, 5-HT3C2, A1BG, A1BG-AS1, A1CF, A2M, A2M-AS1 
ref_counts="resources/cell_refs/allen_institute/allen_m1c_2019_ssv4.rds"
ref_motor_cortex <- readRDS(ref_counts)

ref_motor_cortex

# head(ref_motor_cortex@meta.data)
```
Yadav:

fastq files were run through 10x Genomics Cellranger v3.0.2
output bam files from cellranger were processed using BAMboozle for anonymization
* Genome_build: hg38
Supplementary_files_format_and_content: Counts tables for all barcodes in standard hdf5 output from Cellranger

* https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSM5723843

```{r}
ref_counts_sc=data.table::fread("resources/cell_refs/Yadav_2023/GSE190442_aggregated_counts_postqc.csv")
ref_counts_sc <- as.data.frame(ref_counts_sc)
rownames(ref_counts_sc) <- ref_counts_sc$V1
ref_mat_sc <- as.matrix(ref_counts_sc[,-1])
head(ref_mat_sc[,1:5])
dim(ref_mat_sc) #22238 55290

ref_cell_meta=read.csv("resources/cell_refs/Yadav_2023/GSE190442_aggregated_metadata_postqc.csv", row.names = 1)
dim(ref_cell_meta) # 55289     6

ref_spinal_cord <- CreateSeuratObject(counts = ref_mat_sc,
                                   meta.data = ref_cell_meta,
                                   assay="RNA",
                                   project="yadav_2023")

ref_spinal_cord
# saveRDS(ref_spinal_cord, "resources/cell_refs/Yadav_2023/GSE190442_aggregated_postqc_seurat_obj.RDS")
```

```{r}
head(ref_motor_cortex@assays$RNA)
head(ref_spinal_cord@assays$RNA)

genes_common <- intersect(rownames(ref_motor_cortex), rownames(ref_spinal_cord))
length(genes_common) #20243 only in common :/
```

```{r}
missing_genes <- data.frame(gene_name=rownames(ref_motor_cortex),
           data_set="mc") %>% 
  full_join(., data.frame(gene_name=rownames(ref_spinal_cord),
                          data_set="sc"),
            by="gene_name")

missing_genes %>% 
  filter(is.na(data_set.x))
```


# Session Info

```{r}
session_info()
```

