---
title: "iPSC Motor Neurons Time-series Data: Trajectory Analysis"
author: "Jenny L Smith"
date: "`R Sys.Date()`"
output: html_document
---

# Set-up 

```{r set-up, echo=FALSE}
knitr::opts_knit$set(root.dir = rprojroot::find_rstudio_root_file())
knitr::opts_chunk$set(tidy.opts=list(width.cutoff=80),
                      tidy=TRUE,
                      fig.align='center',
                      fig.width = 10, fig.height = 10,
                      eval = TRUE)

options(stringsAsFactors = FALSE, max.print = 100)
options(Seurat.object.assay.version = 'v5')
table = function (..., useNA = 'ifany') base::table(..., useNA = useNA)
getwd()
```

```{r message=FALSE,warning=FALSE}
library(dplyr)
library(tidyr)
library(magrittr)
library(stringr)
library(tibble)
library(glue)

library(ggplot2)
library(plotly)
library(RColorBrewer)

library(Seurat)
# library(SeuratDisk)
# library(SeuratObject)
# library(SeuratWrappers)
# library(Matrix)
# library(SeuratObject)
# library(SeuratData)
# library(SeuratWrappers)

library(scDblFinder)
library(SoupX)
```

```{r}
conflicted::conflict_prefer("select", "dplyr")
conflicted::conflict_prefer("filter","dplyr")
conflicted::conflict_prefer("saveRDS","base")
```

# Define Functions

```{r}
colors_vector <- c(RColorBrewer::brewer.pal(12, "Paired"), 
            "blue1", "darkslategray3", "burlywood3", "#984EA3",
            "seagreen1", "yellow2", "orchid", "darkblue", 
            "lightsalmon2","slateblue1","lightskyblue4",
            "azure2", "chartreuse1",  "lemonchiffon2",
            "deeppink", "darkslategray1", "green4", "navajowhite", 
            "brown4", "darkgoldenrod2", "deepskyblue1", "lightpink") %>% 
  # c(., ggpubr::get_palette("jco", 5)) %>% 
  c("mediumpurple4","magenta", "peru") %>% 
  c(ggsci::pal_igv()(51))

length(colors_vector) 

sample_id_colors <- c("black",colors_vector[1:18])
```

```{r}
make_3D_umap <- function(metadata, idents_column, label_cols, selected_colors=NULL){
    "
    idents_column is for labeling. ex) 'RNA_snn_res.0.7 as well as the UMAP colnames, eg UMAP_1, UMAP_2'
    "
    # 3D UMAP
    plot.data <- metadata
    plot.data$label <- paste(rownames(plot.data))

    plot.data  <- plot.data %>% 
          unite(col = label, all_of(c("label",label_cols)), sep = "\n", remove = FALSE) %>% 
          select(all_of(c("label",label_cols)), all_of(idents_column))

    
    if(is.null(selected_colors)){
      selected_colors <- c("lightseagreen", "gray50",  "darkgreen", "red4", "red", "turquoise4","black","yellow4","royalblue1","lightcyan3",
                 "peachpuff3","lightskyblue","gray20","orange2","royalblue4","yellow3","gray80","darkorchid1","lawngreen","plum2", 
                 "darkorange4","darkmagenta","mediumorchid1","lightslateblue","midnightblue", "tan2") #26
    }

    cnames <- colnames(plot.data)
    formulas <- sapply(1:3, function(x) as.formula(paste0("~", grep(glue("_{x}$"), cnames, value = TRUE))))
    fig <- plot_ly(data = plot.data, 
        x = formulas[[1]], y = formulas[[2]], z = formulas[[3]], 
        color = as.formula(paste0("~", label_cols[1])), 
        colors = selected_colors, 
        type = "scatter3d", 
        mode = "markers", 
        marker = list(size = 1, width=2), # controls size of points
        text=~label, # This is that extra column we made earlier for which we will use for cell ID
        hoverinfo="text") #When you visualize your plotly object, hovering your mouse pointer over a point shows cell names

    return(fig)
}
```

```{r}
custom_dimplot <- function(metadata, colorby, dimred_colnames, custom_colors,
                           lab_size = 6, ny = 0, custom_lab_fill = NULL,
                           label_samples = TRUE,
                           color_text_grp = FALSE,
                           add_text = TRUE, 
                           seed = 2023){
  
  # define x and y column names
  x_names <- paste0(dimred_colnames, "_1")
  y_names <- paste0(dimred_colnames, "_2")
  
  # define labels positions
  labs_df <- metadata %>%
    select(all_of(c(colorby, x_names, y_names))) %>%
    group_by(!! as.name(colorby)) %>% 
    summarise(x = mean(!! as.name(x_names)),
              y = mean(!! as.name(y_names)))
  
   # define point sizes based on how many total cells in data
   ncells <- nrow(metadata)
   pt_size <- case_when(
     ncells < 5e4 ~ 1,
     ncells >= 5e4 & ncells < 1e5 ~ 0.5,
     ncells >= 1e5 ~ 0.25)
   
   # Make the scatterplot
   plot <- ggplot(metadata, 
         aes_string(x = x_names, y = y_names)) +
    geom_point(aes_string(color = colorby),
               size = pt_size, alpha = 0.4) +
    scale_color_manual(values = custom_colors) +
    guides(color = guide_legend(override.aes = list(alpha = 1, size = 2))) +
    theme_classic() +
    theme(legend.position = "top")
   
   # customize label positions
   if(is.character(ny)){
     y_val <-  ifelse(grepl(ny, labs_df[[colorby]]), 1.5, 0)
   }else{
     y_val <- ny
   }
   
   if(is.null(custom_lab_fill)){
     fill_val <-  "white"
   }else{
     fill_val <- custom_lab_fill
   }
   
   if(color_text_grp){
     mapping <- aes_string(x = "x", y = "y", label = colorby, color = colorby)

   }else{
     mapping <- aes_string(x = "x", y = "y", label = colorby)
   }
  
   if( all(add_text & label_samples) ){
     plot <- plot + 
           ggrepel::geom_text_repel(data = labs_df, 
                             mapping = aes_string(x = "x", y = "y", label = colorby),
                             color = "black",
                             nudge_y = y_val,
                             segment.colour = NA,
                             size = lab_size,
                             seed = seed)
   }else if ( label_samples ) {
      plot <- plot + 
         ggrepel::geom_label_repel(data = labs_df, 
                           mapping = mapping,
                           fill = fill_val,
                           nudge_y = y_val,
                           segment.colour = NA,
                           label.size = 0.5,
                           size = lab_size,
                           show.legend = FALSE,
                           seed = seed)
   }
   
   return(plot)
}
```

```{r}
custom_heatmaps <- function(seurat_obj, genes,columns = NULL, annot_colors = NULL, celltype="ALL"){
  
  if(celltype == "ALL"){
    cell_ids <- colnames(seurat_obj)
  }else{
    cell_ids <- CellsByIdentities(seurat_obj, idents = celltype) %>%  unlist()
  }
  
  print(c("The assay is", DefaultAssay(seurat_obj)))
  genes <- intersect(genes,rownames(seurat_obj))
  seurat_obj <- ScaleData(seurat_obj, features = genes)
  
  print("Extracting expression data")
  norm_counts <- LayerData(seurat_obj[[DefaultAssay(seurat_obj)]], 
                           layer = 'scale.data', 
                           features = genes, 
                           cells = cell_ids)
  
  col_annots <- seurat_obj@meta.data %>% 
    select(-all_of("barcode")) %>% 
    rownames_to_column("barcode") %>% 
    filter(barcode %in% colnames(norm_counts)) %>% 
    as.data.frame() %>%
    set_rownames(.$barcode) %>%
    select(all_of(columns))
  
  # row_annots <- temp %>% 
  #   set_rownames(.$coords) %>% 
  #   select(gene_name, p_val, p_val_adj) 
  
  x <- nrow(norm_counts)
  fontsize_rows <- case_when(
    x < 50 ~ 10,
    x >= 50 & x < 100 ~ 8, 
    x >= 100 & x < 200 ~ 6, 
    x >= 200 ~ 4)


  len <- 299
  colors <- colorRampPalette(viridis::viridis(10))(len)
  colors_list <- lapply(columns, function(x){
          grps <- unique(pull(col_annots, all_of(x)))
          n <- length(grps)
          colors <- annot_colors[1:n] %>% 
            set_names(grps)
          return(colors)
    })
  names(colors_list) <- columns
  

  pheatmap::pheatmap(mat=as.matrix(norm_counts), 
                     scale = "none", 
                     color = colors,
                     fontsize_row = fontsize_rows,
                     # main = glue::glue("{celltype}: DA Peaks p<0.05 & FC >= {th}"),
                     annotation_col = col_annots,
                     annotation_colors = colors_list,
                     # annotation_row = row_annots[,-1],
                     labels_row = rownames(norm_counts),
                     clustering_method = 'ward.D2',
                     show_colnames = FALSE)
}
```



# Parallelization 

```{r eval=FALSE}
library(future)
# check the current active plan
plan()
plan("multisession", workers = 16)
plan()
options(future.globals.maxSize = 60000 * 1024^2)
```

# Genome Refs

```{r eval=FALSE}
suppressPackageStartupMessages(library(ensembldb))

# db <- ensDbFromGtf("resources/genome/Homo_sapiens.GRCh38.109.chr.gtf.gz")

#Warning: Could not determine length for all seqnames
db <- ensDbFromGtf(gtf = "/gpfs/shared_data/10X/cellranger-7.0.0/refdata-gex-GRCh38-2020-A/genes/genes.gtf", 
                   genomeVersion = "GRCh38",
                   organism = "Homo_sapiens",
                   version = 98)
```

```{r, eval=FALSE}
edb <- ensembldb::EnsDb("resources/genome/Homo_sapiens.GRCh38.98.sqlite")

# edb

gids <- ensembldb::keys(edb, keytype = "GENEID")
idmap <- AnnotationDbi::select(edb, keytype="GENEID", 
                      keys = gids, 
                      columns = c("GENEID","GENEBIOTYPE","GENENAME","SYMBOL")) %>% 
  janitor::clean_names()

# head(idmap)
# dim(idmap) #36601
```

```{r}
changho_refs_file <- "references/changho_updated_marker_genes_MN_IN_annotation.xlsx"
marker_genes_chun <- purrr::map_dfr(openxlsx::getSheetNames(changho_refs_file)[2:4], function(x){
  openxlsx::read.xlsx(changho_refs_file, sheet = x)
}) %>% 
  janitor::clean_names() %>% 
  distinct() %>% 
  mutate(group=paste(author, tolower(neural_pop)) %>% 
           gsub("\\s","_",.)) %>% 
  mutate(specific_markers_for_each_type_of_neuron=gsub("\\s","", specific_markers_for_each_type_of_neuron))

# marker_genes_chun
marker_genes_chun_list <- purrr::map(unique(marker_genes_chun$group), function(x){
  marker_genes_chun %>%
    filter(group == x) %>%
    pull(specific_markers_for_each_type_of_neuron)
}) %>% 
  set_names(unique(marker_genes_chun$group))
marker_genes_chun_list[["pan_neuronal_markers"]] <- str_split(marker_genes_chun[1,"pan_neuronal_markers"], pattern = ", ", n=3)[[1]]

# marker_genes_chun_list
```

```{r}
marker_genes_endoderm <- c("CXCR4", "HDE1", "HDE2", "SOX17", "KLF8", "MYCT1", "DKK4",
                           "CLDN6","KRT19","EOMES", "FABP1", "FABP2","GATA4","GSC",
                           "KRT12","FOXA1","FOXA2","TGFB1","SOX7","SOX17", "HNF1B")

length(marker_genes_endoderm)
```

```{r}
# https://github.com/regan-hamel/h-iPSCs-MNI/blob/master/R%20workflow_iMNs_LThiry_2020.Rmd
marker_genes_iPSC <- list(MNPCs = c("CHAT", "ISL1", "ISL2", "MNX1", "OLIG2", "NEUROG2"),
                INs = c("PAX3", "VSX2", "GATA3", "SOX14", "SIM1", "LBX1", "TLX3"),
                NPCs = c("SOX1", "SOX2", "MKI67"),
                Glial = c("SOX9", "S100B", "GFAP"),
                Oligodendrocytes = c("PDGFRA", "GALC"),
                iPSCs = c("NANOG", "POU5F1"))
```

```{r warning=FALSE}
rayon_marker_genes <- openxlsx::read.xlsx(xlsxFile = "resources/cell_refs/Rayon_2021/TableS2.xlsx") %>% 
  janitor::clean_names() %>% 
  mutate(id = paste("rayon", type, neural_pop, sep="_") %>% 
           gsub("\\s", "_", .)) %>% 
  mutate(description = case_when(
    grepl("FP", id) ~ gsub("FP","floor plate", id),
    grepl("RP", id) ~ gsub("RP","roof plate", id),
    grepl("dp", id) ~ gsub("dp[0-9]", "dorsal interneuron progenitors", id),
    grepl("p[0-2]", id) ~ gsub("p[0-2]", "intermediate interneuron progenitors", id),
    grepl("p3", id) ~ gsub("p3","ventral interneuron progenitors p3", id), 
    grepl("V[0-3]", id) ~ gsub("V[0-9].?", "ventral neurons", id),
    grepl("MN", id) ~ gsub(".?MN","motor neuron", id),
    grepl("dl[0-6]", id) ~ gsub("dl[0-9]", "dorsal Interneuron", id),
    grepl("LTMR|LMTR", id) ~ gsub("C?-{0,1}(LTMR|LMTR).{0,}", "mechanoreceptor", id),
    grepl("Proprioceptor|nociceptor", id) ~ gsub("Peripheral","PNS Peripheral",id),
    grepl("Mesoderm|Eryt|Blood|Hema|Myoblast|Oligo", id) ~ paste("rayon",str_split_fixed(id, "_", n=3)[,2]),
    TRUE ~ id
  ) %>% gsub("_"," ", .) %>% gsub("rayon ","", .))


# View(rayon_marker_genes)
# head(rayon_marker_genes)

rayon_marker_genes_long <- rayon_marker_genes %>% 
  separate(genes_map_step1, into = paste0("gene",1:5), sep=", ") %>% 
  separate(genes_map_step2, into = paste0("gene",5:20), sep = ", ") %>% 
  pivot_longer(cols = matches("^gene"),
               names_to = "colname",
               values_to = "genes") %>% 
  filter(!is.na(genes)) 

# rayon_marker_genes_long

rayon_marker_gene_list <-  purrr::map(unique(rayon_marker_genes_long$id), function(x){
  rayon_marker_genes_long %>% 
    filter(id == x) %>% 
    pull(genes) %>% 
    unique()
})
names(rayon_marker_gene_list) <- unique(rayon_marker_genes_long$id)
# rayon_marker_gene_list
```

```{r}
rayon_degs <- openxlsx::read.xlsx("resources/cell_refs/Yadav_2023/1-s2.0-S0896627323000314-mmc5.xlsx", sheet = 1, check.names = FALSE, sep.names = " ") %>% 
  pivot_longer(cols = everything(),
               names_to = "subtype_annotation", 
               values_to = "degs") %>% 
  mutate(names = gsub("-[0-9]{1,2}$","", subtype_annotation) %>% 
           gsub(".+\\s(.+)$", "\\1", .) %>% 
           gsub("\\/|-","_", .))

esc_celltypes <- unique(rayon_degs$names)
rayon_deg_gene_list <-  purrr::map(esc_celltypes, function(x){
  rayon_degs %>%
    filter(names == x) %>%
    pull(degs) %>%
    unique()
})
names(rayon_deg_gene_list) <- esc_celltypes
```

```{r}
mn_subtype_markers <- openxlsx::read.xlsx("references/11122023_Alpha_Gamma_MNs_consolidated marker list.xlsx", sheet = 2)
head(mn_subtype_markers)
# table(mn_subtype_markers$celltype)
```


# Read in iPSC Object

```{r}
current_path <- file.path("results/day0_day50/seurat")
current_obj <- "iPSC_motor_neuron_day0-day50_time-course_filtered_merged_soupX_normalized_dimred_anno_seurat_obj_v5.RDS" #v5 Seurat Object

ipsc_obj <- readRDS(file.path(current_path, current_obj))
ipsc_obj@meta.data$log2_nCount_RNA <- log2(ipsc_obj@meta.data$nCount_RNA)

ipsc_obj
```

```{r}
# logNormalize to be consistent with the reference dataset
DefaultAssay(ipsc_obj) <- 'RNA'

# split to normalize each individually
ipsc_obj[["RNA"]] <- split(ipsc_obj[["RNA"]],
                                  f = ipsc_obj$orig.ident)
ipsc_obj <- NormalizeData(ipsc_obj,
                                 normalization.method = "LogNormalize") %>% 
  FindVariableFeatures() %>% 
  ScaleData() %>% 
  RunPCA(npcs = 50, 
         reduction.name = "pca_lognorm")

# join layers to export the full raw counts matrix for all samples (layers)
ipsc_obj[["RNA"]] <- JoinLayers(ipsc_obj[["RNA"]])
```

```{r}
# find clusters with harmony integrated UMAP
DefaultAssay(ipsc_obj) <- "SCT"

ipsc_obj <- ipsc_obj %>% 
  FindNeighbors(reduction = "harmony",
                dims = 1:50, 
                k.param = 20, #default
                graph.name = c("harmony_nn","harmony_snn")) %>% 
  FindClusters(resolution = 1, 
               algorithm = 3, 
               cluster.name = "harmony_clusters",
               group.singletons = TRUE)
```

```{r}
table(ipsc_obj$harmony_clusters)
#29 clusters
```

## Visualization 

```{r}
table(ipsc_obj$subtype_standardized)
```

```{r fig.height=10, fig.width=10}
custom_dimplot(metadata = ipsc_obj@meta.data, 
               colorby = "batch", 
               dimred_colnames = "umapharmony", 
               custom_colors = ggpubr::get_palette("jco", 4))

custom_dimplot(metadata = ipsc_obj@meta.data, 
               colorby = "subtype_standardized", 
               dimred_colnames = "umapharmony", 
               custom_colors = ggpubr::get_palette("simpsons", 7))

custom_dimplot(metadata = ipsc_obj@meta.data, 
               colorby = "harmony_clusters", 
               dimred_colnames = "umapharmony", 
               custom_colors = ggpubr::get_palette("simpsons", 29))
```

## Investigate Outliers 

```{r}
outliers <- subset(ipsc_obj, subset = umapharmony_2 > 8)

outliers
table(outliers$harmony_clusters)[table(outliers$harmony_clusters) != 0]
```

```{r fig.height=10, fig.width=10}
custom_dimplot(metadata = outliers@meta.data, 
               colorby = "batch",label_samples = FALSE,
               dimred_colnames = "umapharmony", 
               custom_colors = ggpubr::get_palette("jco",4))
```

```{r}
quantile(ipsc_obj$log2_nCount_RNA, probs = seq(0.5,1, length.out = 11))
FeaturePlot(ipsc_obj,
            reduction = "umap.harmony", 
            cols = c("azure","red"),
            max.cutoff = 14.87302,
            features = "log2_nCount_RNA")

FeaturePlot(outliers,
            reduction = "umap.harmony",
            max.cutoff = 14.87302,
            features = "log2_nCount_RNA")

FeaturePlot(ipsc_obj,
            reduction = "umap.harmony", 
            cols = c("azure","red"),
            features = "nFeature_RNA")

FeaturePlot(outliers,
            reduction = "umap.harmony",
            features = "nFeature_RNA")

FeaturePlot(ipsc_obj,
            reduction = "umap.harmony", 
            cols = c("azure","red"),
            features = "percent.mt")
FeaturePlot(outliers,
            reduction = "umap.harmony",
            features = "percent.mt")
```

```{r}
table(table(outliers$orig.ident)  == 1 )
```

```{r fig.height=10}
# outliers[["RNA"]] <- split(outliers[["RNA"]], f = outliers$orig.ident)
umi_counts_c20 <-  sapply(grep("counts",SeuratObject::Layers(outliers@assays$RNA), value = TRUE), function(x) {
   colSums(outliers@assays$RNA[[x]] > 0)
})

sapply(umi_counts_c20, quantile) %>% 
  as.data.frame() %>% 
  pheatmap::pheatmap(mat = ., cluster_rows = FALSE)
```

# Subset (?)

removed 586 cells total 

```{r}
# remove cluster 20 outliers for now. Need to determine what is driving their VERY different expression profile. But the cells looks so clustered so close that I also think it could be technical artifacts that could be making them cluster like that. 
ipsc_obj_subset <- subset(ipsc_obj, subset =  umapharmony_2 < 8)
DefaultAssay(ipsc_obj_subset) <- "SCT"
ipsc_obj_subset
```

```{r}
# re-run find neighbors, UMAP reduction, and clustering on the subset object
dims <- 1:30
ipsc_obj_subset <- ipsc_obj_subset %>% 
    FindNeighbors(reduction = "harmony",
                dims = dims, 
                k.param = 20, #default
                ) %>% 
    FindClusters(resolution = 0.7, 
                 algorithm = 3, 
                 cluster.name = "subset_harmony_clusters",
                 group.singletons = FALSE
                 ) %>% 
     RunUMAP(reduction = "harmony",
           n.components = 3,
           dims = dims,
           n.neighbors = 30,# default 30
           metric = "cosine",
           min.dist = 0.1, # default 0.3
           # negative.sample.rate = 10,
           n.epochs = 500,
           reduction.name = "umap_harmony_subset")

# ipsc_obj
ipsc_obj_subset <- AddMetaData(ipsc_obj_subset, 
                        metadata = ipsc_obj_subset@reductions$umap_harmony_subset@cell.embeddings)

table(ipsc_obj_subset$subset_harmony_clusters) # 22 clusters 
```



```{r fig.height=10, fig.width=10}
custom_dimplot(metadata = ipsc_obj_subset@meta.data, 
               colorby = "batch", 
               dimred_colnames = "umapharmonysubset", 
               custom_colors = ggpubr::get_palette("jco", 4))

custom_dimplot(metadata = ipsc_obj_subset@meta.data, 
               colorby = "subtype_standardized", 
               add_text = FALSE,
               dimred_colnames = "umapharmonysubset", 
               custom_colors = brewer.pal(7,"Set3"))

custom_dimplot(metadata = ipsc_obj_subset@meta.data, 
               colorby = "subset_harmony_clusters", 
               dimred_colnames = "umapharmonysubset", 
               custom_colors = ggpubr::get_palette("simpsons", 22))

custom_dimplot(metadata = ipsc_obj_subset@meta.data, 
               colorby = "id",
               label_samples = FALSE,
               dimred_colnames = "umapharmonysubset", 
               custom_colors = viridis::turbo(19))
```

# Trajectories

Review article:
https://www.nature.com/articles/s41576-021-00444-7

Selecting which time-points to include:
However, there are cases where the dynamics of the process being studied is unknown or when assumptions are based on the phenotypic behaviour, which may not reflect the underlying molecular dynamics that are being profiled by scRNA-seq71. To address this, Kleyman et al. developed the time point selection (TPS) method71 (Fig. 2). TPS was originally developed and applied to bulk RNA-seq but can also be used for scRNA-seq. It works by initially oversampling bulk-level RNA-seq using cheap array methods. Next, spline curves are used to fit the profiled data, which enables the method to predict values for unobserved time points. A heuristic optimization function is then used to select the most informative time points, those points that if sampled provide enough information to reconstruct the entire expression trajectory for all genes


OK - slingshot
   - Monocle3
   - CellRank
No
- PRESCIENT (8 citations, not maintained in 3 yrs)
- psupertime (not maintained in 4yrs, only 11 citations)
- dyno (not maintained in 4yrs)
- CSHMM (not maintained in 7 years, 151 citations)
https://github.com/gifford-lab/prescient

https://github.com/dynverse/dyno

# Monocle3

```{r}
library(monocle3)
library(SeuratWrappers)
```

```{r}
cds <- new_cell_data_set(expression_data = ipsc_obj@assays$SCT$counts,
                         cell_metadata = ipsc_obj@meta.data)
logcounts(cds) <- ipsc_obj@assays$SCT$data

reducedDims(cds) <- list(PCA=ipsc_obj@reductions$pca@cell.embeddings,
                         Harmony=ipsc_obj@reductions$harmony@cell.embeddings,
                         UMAP=ipsc_obj@reductions$umap.harmony@cell.embeddings)

cds
```

```{r}
# use the harmony UMAP (renamed to simply UMAP for compatibility) to identify unsupervised clusters
# defines the 'partitions' found in the cell clustering 
cds <- cluster_cells(cds, 
                     reduction_method = "UMAP",
                     cluster_method = "leiden",
                     k = 20)

cds
```

```{r fig.height=10}
p1 <- plot_cells(cds, 
                 show_trajectory_graph = FALSE)

p2 <- plot_cells(cds, 
                 color_cells_by = "partition", 
                 show_trajectory_graph = FALSE)

p3 <- plot_cells(cds, 
                 color_cells_by = "orig.ident", 
                 show_trajectory_graph = FALSE)

p1
p2
p3
```

```{r}
# Identifying the program of gene expression changes is equivalent to learning a trajectory that the cells follow through this space.
cds <- learn_graph(cds, 
                   use_partition = FALSE,
                   verbose = TRUE)
```

```{r}
DimPlot(subset(ipsc_obj, subset = orig.ident == "iPSC_ipsc"),
        reduction = "umap.harmony",
        group.by = "harmony_clusters")
```

```{r fig.height=10, fig.width=10}
plot_cells(cds, label_principal_points = TRUE)
```

If `root_cells` is set as NULL, then its based on manual selection of the root cells. However, there are methods available like time point selection (TPS) that might be able to refine the root cells, rather than use all time0 cells.

```{r}
#Assigns cells a pseudotime value based on their projection on the principal graph
time0 <- WhichCells(ipsc_obj, expression = orig.ident == "iPSC_ipsc")
cds <- order_cells(cds, 
                   reduction_method = "UMAP", 
                   root_cells = time0, 
                   verbose = TRUE)

# cds
```

```{r fig.height=10, fig.width=10}
plot_cells(cds, 
           label_groups_by_cluster = FALSE,
           label_leaves = FALSE, 
           label_branch_points = FALSE)
```

```{r fig.height=10, fig.width=10}
plot_cells(cds, 
           color_cells_by = "pseudotime",
           label_groups_by_cluster = FALSE,
           label_leaves = FALSE, 
           label_branch_points = FALSE)
```

# Slingshot

```{r}
library(slingshot)
```

cds <- new_cell_data_set(expression_data = ipsc_obj@assays$SCT$counts,
                         cell_metadata = ipsc_obj@meta.data)
logcounts(cds) <- ipsc_obj@assays$SCT$data

reducedDims(cds) <- list(PCA=ipsc_obj@reductions$pca@cell.embeddings,
                         Harmony=ipsc_obj@reductions$harmony@cell.embeddings,
                         UMAP=ipsc_obj@reductions$umap_harmony_subset@cell.embeddings)

two-step process composed of identifying the global lineage structure with a cluster-based minimum spanning tree (MST) and fitting simultaneous principal curves to describe each lineage.

```{r}
sce <- Seurat::as.SingleCellExperiment(ipsc_obj, assay = "SCT")

sce
```

potential labels
1) subtype (N=7)
2) unsupervised clusters (N=22)
* subset_harmony_clusters
3) subtype + unsupervised clusters

```{r}
sce$subtype_standardized_subclusters %>% table()
reducedDims(sce)
head(reducedDim(sce, "UMAP.HARMONY"))
```

All of the results are stored in a PseudotimeOrdering object, which is added to the colData of the original object and can be accessed via colData(sce)$slingshot

to extact all slingshot objects/results 
  * as.SlingshotDataSet (class SlingshotDataSet) 
  * as.PseudotimeOrdering (classSummarizedExperiment)


omega
This value is provided as a potentially useful rule of thumb for datasets with outlying clusters or multiple, distinct trajectories. 

 recommend the **specification of an initial cluster based on prior knowledge**
  - either time of sample collection or established gene markers
  - This specification will have no effect on how the MST is constructed, but it will impact how branching curves are constructed.

```{r}
sce <- slingshot(sce, 
                 clusterLabels = 'subtype_standardized_subclusters',
                 reducedDim = 'HARMONY',
                 time = sce$time_point,
                 start.clus = NULL,
                 end.clus = NULL,
                 dist.method = "slingshot",
                 use.median = FALSE, 
                 reassign = TRUE, 
                 allow.breaks = TRUE)
```

```{r}
slingshot_data <- SlingshotDataSet(sce)

# slingshot_data
# sce
slingshot_data@lineages %>% head()
```


```{r}
sce2 <- slingshot(sce, 
                 clusterLabels = 'subtype_standardized_subclusters',
                 reducedDim = 'UMAP.HARMONY',
                 time = sce$time_point,
                 start.clus = NULL,
                 end.clus = NULL,
                 dist.method = "slingshot",
                 use.median = FALSE, 
                 reassign = TRUE, 
                 allow.breaks = TRUE)
```

colors_vector
```{r fig.height=10, fig.width=10}
plot(reducedDim(sce, "UMAP.HARMONY"), 
     col = colors_vector[sce$subtype_standardized_subclusters],
     asp = 1, pch = 16)
lines(SlingshotDataSet(sce2),
      type = 'lineages',
      lwd = 2, 
      col = 'black')
```




# TSCAN

```{r}
library(TSCAN)
```

# CellRank 

https://cellrank.readthedocs.io/en/latest/notebooks/tutorials/kernels/500_real_time.html

a scRNA-seq dataset with experimental time points. Additional spatial or lineage information improves the accuracy of fate mapping, but is not required.

# Session Info

```{r}
sessionInfo()
```

