---
title: "iPSC Motor Neurons and Spinal Cord Cell Refs Data"
author: "Jenny L Smith"
date: "`R Sys.Date()`"
output: html_document
---

# Set-up 

```{r set-up, echo=FALSE}
knitr::opts_knit$set(root.dir = rprojroot::find_rstudio_root_file())
knitr::opts_chunk$set(tidy.opts=list(width.cutoff=80),
                      tidy=TRUE,
                      fig.align='center',
                      fig.width = 10, fig.height = 10,
                      eval = TRUE)

options(stringsAsFactors = FALSE, max.print = 100)
options(Seurat.object.assay.version = "v5")
table = function (..., useNA = 'ifany') base::table(..., useNA = useNA)
```

```{r message=FALSE,warning=FALSE}
library(dplyr)
library(tidyr)
library(magrittr)
library(stringr)
library(tibble)
library(glue)

library(ggplot2)
library(plotly)
library(RColorBrewer)

library(Seurat)
```

```{r}
conflicted::conflict_prefer("select", "dplyr")
conflicted::conflict_prefer("filter","dplyr")
conflicted::conflict_prefer("saveRDS","base")
```


# Parallelization 

```{r}
library(future)
# check the current active plan
plan()

# plan("multisession", workers = 2)
# plan()
options(future.globals.maxSize = 10000 * 1024^2)
```


# Reference Data Set(s)

1. https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE136719
2. https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE138121


3.  Rayon 2021  https://github.com/briscoelab/human_single_cell
3-1 https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE171892
3-2. https://shiny.crick.ac.uk/scviewer/neuraltube/

4. https://juliendelile.github.io/Antler/


5. Yadav 2023 
  https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE190442
  https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE222322



```{bash eval=FALSE}
wget https://ftp.ncbi.nlm.nih.gov/geo/series/GSE190nnn/GSE190442/suppl/GSE190442_aggregated_counts_postqc.csv.gz
wget https://ftp.ncbi.nlm.nih.gov/geo/series/GSE190nnn/GSE190442/suppl/GSE190442_aggregated_metadata_postqc.csv.gz

rclone copy --http-url https://zenodo.org :http:record/4546932 . -P
rclone copy --http-url http://celltypes.brain-map.org :http:api/v2/well_known_file_download/694416044 . -P
rclone copy --http-url http://celltypes.brain-map.org :http:api/v2/well_known_file_download/502175284 . -P
rclone copy refseq:/geo/series/GSE190nnn/GSE190442/suppl/GSE190442_RAW.tar . -P
```

```{bash eval=FALSE}
cd resources/cell_refs/allen_institute/M1

# https://azimuth.hubmapconsortium.org/references/#Human%20-%20Motor%20Cortex
#wget https://seurat.nygenome.org/azimuth/demo_datasets/allen_m1c_2019_ssv4.rds

rclone copy --http-url https://brainmapportal-live-4cc80a57cd6e400d854-f7fdcae.divio-media.net :http:filer_public/70/32/70326830-e306-4743-a02c-a8da5bf9eb56/readme-m1-10.txt  . -P

rclone copy --http-url https://brainmapportal-live-4cc80a57cd6e400d854-f7fdcae.divio-media.net \
  :http:filer_public/0c/0c/0c0c882d-1c31-40a9-8039-3bf2706a77cd/sample-exp_component_mapping_human_10x_apr2020.zip . -P 

rclone copy --http-url https://idk-etl-prod-download-bucket.s3.amazonaws.com :http:aibs_human_m1_10x/metadata.csv . -P
rclone copy --http-url https://idk-etl-prod-download-bucket.s3.amazonaws.com :http:aibs_human_m1_10x/matrix.csv . -P
```


```{bash eval=FALSE}
# https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE171892
cd resources/cell_refs/Rayon_2021
rclone copy refseq:/geo/series/GSE171nnn/GSE171892/suppl . -P

for file in $(ls -1 *.gz)
do
  tar -xzvf $file 
done

cd resources
git clone https://github.com/briscoelab/human_single_cell
```


# Genome Refs

```{r eval=FALSE}
edb <- ensembldb::EnsDb("resources/genome/Homo_sapiens.GRCh38.98.sqlite")

# edb

gids <- ensembldb::keys(edb, keytype = "GENEID")

idmap <- AnnotationDbi::select(edb, keytype="GENEID", 
                      keys = gids, 
                      columns = c("GENEID","GENEBIOTYPE","GENENAME","SYMBOL")) %>% 
  janitor::clean_names()

head(idmap)
```

```{r eval=FALSE}
MTG_genes <- read.csv("resources/cell_refs/allen_institute/MTG/human_MTG_2018-06-14_genes-rows.csv")

head(MTG_genes)
```

```{r}
marker_genes_mc <- read.csv("resources/cell_refs/allen_institute/allen_institute_azimuth_marker_gene_list.csv") %>% 
  mutate(Label = gsub("\\s{1,}|\\/|-","_", Label)) %>% 
  pull(Markers , name = Label) %>% 
  as.list() %>% 
  sapply(.,  function(x) str_split(gsub("\\.","-",x), pattern = ", "))
```

```{r}
changho_refs_file <- "references/changho_updated_marker_genes_MN_IN_annotation.xlsx"
marker_genes_chun <- purrr::map_dfr(openxlsx::getSheetNames(changho_refs_file)[2:4], function(x){
  openxlsx::read.xlsx(changho_refs_file, sheet = x)
}) %>% 
  janitor::clean_names() %>% 
  distinct() %>% 
  mutate(group=paste(author, tolower(neural_pop)) %>% 
           gsub("\\s","_",.)) %>% 
  mutate(specific_markers_for_each_type_of_neuron=gsub("\\s","", specific_markers_for_each_type_of_neuron))

# marker_genes_chun
marker_genes_chun_list <- purrr::map(unique(marker_genes_chun$group), function(x){
  marker_genes_chun %>%
    filter(group == x) %>%
    pull(specific_markers_for_each_type_of_neuron)
}) %>% 
  set_names(unique(marker_genes_chun$group))
marker_genes_chun_list[["pan_neuronal_markers"]] <- str_split(marker_genes_chun[1,"pan_neuronal_markers"],pattern = ", ", n=3)[[1]]

# marker_genes_chun_list
```

```{r}
# https://github.com/regan-hamel/h-iPSCs-MNI/blob/master/R%20workflow_iMNs_LThiry_2020.Rmd
marker_genes_iPSC <- list(MNPCs = c("CHAT", "ISL1", "ISL2", "MNX1", "OLIG2", "NEUROG2"),
                INs = c("PAX3", "VSX2", "GATA3", "SOX14", "SIM1", "LBX1", "TLX3"),
                NPCs = c("SOX1", "SOX2", "MKI67"),
                Glial = c("SOX9", "S100B", "GFAP"),
                Oligodendrocytes = c("PDGFRA", "GALC"),
                iPSCs = c("NANOG", "POU5F1"))
```

```{r}
marker_genes_endoderm <- c("CXCR4", "KRT12", "SOX17", "KLF8", "MYCT1", "DKK4", "GATA4", "GATA6","SOX7", "EOMES","KRT19","CLDN6","FOXA1","FOXA2")

length(marker_genes_endoderm)
```

https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4468382/
Cellular identities in vertebrate spinal cord are specified during development along the three basic spatial axes of the embryonic body plan 
  * rostral–caudal, dorsal–ventral, and medial–lateral.

```{r}
rayon_marker_genes <- openxlsx::read.xlsx(xlsxFile = "resources/cell_refs/Rayon_2021/TableS2.xlsx") %>% 
  janitor::clean_names() %>% 
  mutate(id = paste("rayon", type, neural_pop, sep="_") %>% 
           gsub("\\s", "_", .)) %>% 
  mutate(description = case_when(
    grepl("FP", id) ~ gsub("FP","floor plate", id),
    grepl("RP", id) ~ gsub("RP","roof plate", id),
    grepl("dp", id) ~ gsub("dp[0-9]", "dorsal interneuron progenitors", id),
    grepl("p[0-2]", id) ~ gsub("p[0-2]", "intermediate interneuron progenitors", id),
    grepl("p3", id) ~ gsub("p3","ventral interneuron progenitors p3", id), 
    grepl("V[0-3]", id) ~ gsub("V[0-9].?", "ventral neurons", id),
    grepl("MN", id) ~ gsub(".?MN","motor neuron", id),
    grepl("dl[0-6]", id) ~ gsub("dl[0-9]", "dorsal Interneuron", id),
    grepl("LTMR|LMTR", id) ~ gsub("C?-{0,1}(LTMR|LMTR).{0,}", "mechanoreceptor", id),
    grepl("Proprioceptor|nociceptor", id) ~ gsub("Peripheral","PNS Peripheral",id),
    grepl("Mesoderm|Eryt|Blood|Hema|Myoblast|Oligo", id) ~ paste("rayon",str_split_fixed(id, "_", n=3)[,2]),
    TRUE ~ id
  ) %>% gsub("_"," ", .) %>% gsub("rayon ","", .))


# View(rayon_marker_genes)
head(rayon_marker_genes)
```

```{r}
rayon_marker_genes_long <- rayon_marker_genes %>% 
  separate(genes_map_step1, into = paste0("gene",1:5), sep=", ") %>% 
  separate(genes_map_step2, into = paste0("gene",5:20), sep = ", ") %>% 
  pivot_longer(cols = matches("^gene"),
               names_to = "colname",
               values_to = "genes") %>% 
  filter(!is.na(genes)) 

# rayon_marker_genes_long

rayon_marker_gene_list <-  purrr::map(unique(rayon_marker_genes_long$id), function(x){
  rayon_marker_genes_long %>% 
    filter(id == x) %>% 
    pull(genes) %>% 
    unique()
})
names(rayon_marker_gene_list) <- unique(rayon_marker_genes_long$id)
# rayon_marker_gene_list
```

```{r}
colors_vector <- c("#E41A1C", "#377EB8", "#4DAF4A", "#F781BF", 
            "blue1", "darkslategray3", "burlywood3", "#984EA3", 
            "#FF7F00", "seagreen1", "maroon", "orchid", "darkblue", 
            "azure2", "chartreuse1", "orange1", 
            "deeppink", "darkslategray1", "green4", "navajowhite2", 
            "brown3", "darkgoldenrod4", "deepskyblue1", "lightcoral") %>% 
  c(., ggpubr::get_palette("jco", 4),
    ggpubr::get_palette("npg", 4)) %>% 
  c("darkorchid1","magenta", "peru") %>% 
  c(ggsci::pal_igv()(51))

length(colors_vector) 
```


# Preprocessing

## Allen Institute:

* This data set includes single-nucleus transcriptomes from 76,533 total nuclei derived from 2 post-mortem human brain specimens, to survey 
cell type diversity in the primary motor cortex (M1C or M1).
*  default 10x Cell Ranger v3


* https://azimuth.hubmapconsortium.org/references/#Human%20-%20Motor%20Cortex
* https://github.com/satijalab/azimuth-references/tree/master/human_motorcortex 
* https://portal.brain-map.org/atlases-and-data/rnaseq/human-m1-10x 
* https://knowledge.brain-map.org/data/8GU8FPSNLBUBCJWQDZ5/summary
* https://www.nature.com/articles/s41586-021-03465-8

Ontologies
* https://bioportal.bioontology.org/ontologies/PCL/?p=summary
* https://data.bioontology.org/ontologies/PCL/submissions/21/download?apikey=8b5b7825-538d-40e0-9e9e-5ab9274a9aeb
* https://academic.oup.com/bioinformatics/article/33/7/1104/2843897

```bash
# http://robot.obolibrary.org/convert
robot convert --input pcl.owl --output pcl.obo --check false
```

Code
* https://github.com/AllenInstitute/BICCN_M1_Evo


```{r}
allen_ontology <- get_ontology("resources/cell_refs/allen_institute/pcl.obo",
                               extract_tags='everything')
# allen_ontology

human_entries <- data.frame(name = allen_ontology$name, 
                   id = allen_ontology$id,
                   def = allen_ontology$def) %>% 
  filter(!grepl("CHEBI|BFO", id), 
         !grepl("obsolete", name)) %>% 
  filter(grepl("PCL", id), grepl("Hsap", name))

```


```{r}
mc_meta <- read.csv("resources/cell_refs/allen_institute/M1/metadata.csv")
rownames(mc_meta) <- mc_meta$sample_name

head(mc_meta)

sample_mapping <- read.csv("resources/cell_refs/allen_institute/M1/sample-exp_component mapping_human_10x_apr2020.csv")

# head(sample_mapping)

az_mc_meta <- read.csv("resources/cell_refs/allen_institute/allen_institute_azimuth_subclass_marker_gene_list.csv") %>% 
  janitor::clean_names()

# head(az_mc_meta)
```

```{r}
celltype_labels <- unique(mc_meta$cluster_label)
length(celltype_labels) #127

defs <- purrr::map_dfr(celltype_labels, function(label){
  human_entries %>% 
    filter(grepl(glue("^{label} "), name), 
           grepl("primary motor cortex", name)) %>% 
    mutate(cluster_label = label)
}) %>% 
  select(cluster_label, everything())

head(defs)
dim(defs)
# celltype_labels[!celltype_labels %in% defs$cluster_label] # missing 2 entries
```

```{r fig.width=10}
cell_tax <- readRDS("resources/cell_refs/allen_institute/M1/human_dendrogram.rds")
plot(cell_tax)
```

```{r}
mc_meta %>% 
  group_by(cluster_label) %>% 
  dplyr::count() %>% 
  arrange(desc(n))

mc_meta %>% 
  group_by(subclass_label) %>% 
  dplyr::count() %>% 
  arrange(desc(n)) %>% 
  write.csv(., "allen_celltypes.csv", row.names = FALSE)
```

```{r eval=FALSE}
ref_mc_counts <- data.table::fread("resources/cell_refs/allen_institute/M1/matrix.csv") 


ref_mc_counts <- as.data.frame(ref_mc_counts)
rownames(ref_mc_counts) <- ref_mc_counts$sample_name
ref_mc_counts <- t(as.matrix(ref_mc_counts[,-1]))

dim(ref_mc_counts) #76533 50282
head(ref_mc_counts[,1:5])
```

```{r eval=FALSE}
ref_motor_cortex <- CreateSeuratObject(counts = ref_mc_counts,
                                   meta.data = mc_meta,
                                   assay="RNA",
                                   project="Allen_M1") 
rm(ref_mc_counts)
rm(mc_meta)
gc()

ref_motor_cortex <- ref_motor_cortex %>% 
  PercentageFeatureSet(
                         assay="RNA",
                         pattern = "^MT-",
                         col.name = "percent.mt") %>% 
  PercentageFeatureSet(
                       assay="RNA",
                       pattern = "^RP[SL]", 
                       col.name = "percent.ribo") %>% 
  SCTransform(
              vst.flavor = "v2", 
              method = "glmGamPoi",
              verbose = T) %>%
    RunPCA(npcs = 50, verbose = T) %>%
    RunUMAP(reduction = "pca", 
            dims = 1:50,
            n.components = 3, 
            min.dist = 0.3,
            verbose = T) %>%
    FindNeighbors(reduction = "pca",
                  dims = 1:50, verbose = T) %>%
    FindClusters(resolution = 0.8, verbose = T)

# saveRDS(ref_motor_cortex,"resources/cell_refs/allen_institute/M1/human_M1_normalized_seurat_object.RDS")
```

```{r}
ref_motor_cortex <- readRDS("resources/cell_refs/allen_institute/M1/human_M1_normalized_seurat_object.RDS")

new_meta <- ref_motor_cortex@meta.data %>% 
  left_join(., az_mc_meta, by=c("cluster_label"="label")) %>% 
  select(sample_name,
         cell_type_designation_label,
         expanded_label, 
         obo_ontology_id,
         cell_type_alias_label, 
         class_label, 
         cluster_label, 
         markers, 
         everything()) %>% 
  set_rownames(.$sample_name)

ref_motor_cortex@meta.data <- new_meta
ref_motor_cortex <- AddMetaData(ref_motor_cortex, metadata = ref_motor_cortex@reductions$umap@cell.embeddings)

# ref_motor_cortex
dim(ref_motor_cortex) #28271 76,533
Idents(ref_motor_cortex) <- ref_motor_cortex$cluster_label
```

```{r eval=FALSE}
ref_motor_cortex@meta.data %>% 
  group_by(obo_ontology_id, expanded_label) %>% 
  dplyr::count() %>% 
  View()

ref_motor_cortex@meta.data %>% 
  group_by(obo_ontology_id, cluster_label) %>% 
  dplyr::count() %>% 
  View()

ref_motor_cortex@meta.data %>% 
  group_by(obo_ontology_id, cell_type_alias_label, class_label) %>% 
  dplyr::count() %>% 
  View()


ref_motor_cortex@meta.data %>% 
  group_by(obo_ontology_id, subclass_label, class_label) %>% 
  dplyr::count() %>% 
  View()
```

```{r}
cluster_color <- ref_motor_cortex@meta.data %>% 
  select(cluster_color, cluster_label) %>% 
  distinct() %>% 
  pull(cluster_color, name = cluster_label)

length(cluster_color)
head(cluster_color)

subclass_color <- ref_motor_cortex@meta.data  %>% 
  select(subclass_label, subclass_color) %>% 
  distinct() %>% 
  pull(subclass_color, name = subclass_label)

table(ref_motor_cortex$external_donor_name_label)
```

```{r fig.height=20, fig.width=20}
# pdf("figures/cell_refs/allen_M1_ref_motor_cortex_cluster_label_UMAP.pdf", height = 20, width = 15)
DimPlot(ref_motor_cortex, 
        repel = TRUE,
        label = TRUE) +
  theme(legend.position = 'bottom') +
  scale_color_manual(values = cluster_color)
# dev.off()
```

```{r fig.height=20, fig.width=20}
# pdf("figures/cell_refs/allen_M1_ref_motor_cortex_subclass_label_UMAP.pdf", height = 20, width = 15)
DimPlot(ref_motor_cortex, 
        group.by = 'subclass_label',
        label.size = 10,
        label.color = 'black',
        label.box = FALSE,
        repel = TRUE,
        label = TRUE) +
  theme(legend.position = 'bottom') +
  scale_color_manual(values = subclass_color)
# dev.off()
```

```{r fig.height=20, fig.width=20}
# pdf("figures/cell_refs/allen_M1_ref_motor_cortex_external_donor_name_label_UMAP.pdf", height = 20, width = 15)
DimPlot(ref_motor_cortex, 
        group.by = "external_donor_name_label",
        label = TRUE) +
  theme(legend.position = 'bottom')
# dev.off()
```

```{r fig.height=20, fig.width=20}
DimPlot(ref_motor_cortex, 
        group.by = "donor_sex_label",
        label = TRUE) +
  theme(legend.position = 'bottom')
```


## Yadav:

fastq files were run through 10x Genomics Cellranger v3.0.2
output bam files from cellranger were processed using BAMboozle for anonymization

* Genome_build: hg38
Supplementary_files_format_and_content: Counts tables for all barcodes in standard hdf5 output from Cellranger

* The 7 human datasets were integrated using SCTransform normalization followed by CCA based integration using the Seurat 

* https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSM5723843

```{r}
sc_pts <- read.csv("resources/cell_refs/Yadav_2023/GSE222322_donor_to_patient_ids.csv")


sc_meta <- read.csv("resources/cell_refs/Yadav_2023/GSE190442_aggregated_metadata_postqc.csv") %>% 
  mutate(barcode=X) %>% 
  left_join(., sc_pts, by=c("sample"="Patient_ID")) %>% 
  column_to_rownames("X") 
  

# head(sc_meta)
dim(sc_meta) # 55289     7

sc_pt_meta <- openxlsx::read.xlsx("resources/cell_refs/Yadav_2023/1-s2.0-S0896627323000314-mmc2.xlsx") %>% 
  janitor::clean_names() %>% 
  filter(use=="Sequencing") %>% 
  mutate(donor = as.character(donor))

# head(sc_pt_meta)
dim(sc_pt_meta)
```

```{r eval=FALSE}
ref_counts_sc <- data.table::fread("resources/cell_refs/Yadav_2023/GSE190442_aggregated_counts_postqc.csv")
ref_counts_sc <- as.data.frame(ref_counts_sc)
rownames(ref_counts_sc) <- ref_counts_sc$V1
ref_mat_sc <- as.matrix(ref_counts_sc[,-1])
head(ref_mat_sc[,1:5])
dim(ref_mat_sc) #22238 55,290

ref_cell_meta=read.csv("resources/cell_refs/Yadav_2023/GSE190442_aggregated_metadata_postqc.csv", 
                       row.names = 1)
dim(ref_cell_meta) # 55289     6

ref_spinal_cord <- CreateSeuratObject(counts = ref_mat_sc,
                                   meta.data = ref_cell_meta,
                                   assay="RNA",
                                   project="yadav_2023")

ref_spinal_cord

# grep("^MT-", rownames(ref_spinal_cord)) #MT genes are removed?
# grep("^RP[SL]", rownames(ref_spinal_cord), value=TRUE)
# saveRDS(ref_spinal_cord, "resources/cell_refs/Yadav_2023/GSE190442_aggregated_postqc_seurat_obj.RDS")

ref_spinal_cord <- readRDS("resources/cell_refs/Yadav_2023/GSE190442_aggregated_postqc_seurat_obj.RDS") %>% 
  PercentageFeatureSet(
                       assay="RNA",
                       pattern = "^MT-",
                       col.name = "percent.mt") %>% 
  PercentageFeatureSet(
                       assay="RNA",
                       pattern = "^RP[SL]", 
                       col.name = "percent.ribo") %>% 
  SCTransform( 
              vst.flavor = "v2", 
              method = "glmGamPoi",
              verbose = T) %>%
    RunPCA(npcs = 50, verbose = T) %>%
    RunUMAP(reduction = "pca", 
            dims = 1:50,
            n.components = 3, 
            min.dist = 0.3,
            verbose = T) %>%
    FindNeighbors(reduction = "pca",
                  dims = 1:50, verbose = T) %>%
    FindClusters(resolution = 0.8, verbose = T)


ref_spinal_cord <- AddMetaData(ref_spinal_cord, select(sc_meta, barcode:Visium_ID))
table(ref_spinal_cord@meta.data$top_level_annotation)
# saveRDS(ref_spinal_cord, "resources/cell_refs/Yadav_2023/GSE190442_aggregated_postqc_normalized_seurat_obj.RDS")
```

```{r}
ref_spinal_cord <- readRDS("resources/cell_refs/Yadav_2023/GSE190442_aggregated_postqc_normalized_seurat_obj.RDS")

sc_new_meta <- sc_meta %>% 
  left_join(., sc_pt_meta, by=c("sample"="donor")) %>% 
  select(barcode, all_of(colnames(.)[colnames(.) %in% colnames(sc_pt_meta)])) %>% 
  column_to_rownames("barcode")

ref_spinal_cord <- AddMetaData(ref_spinal_cord, sc_new_meta)
Idents(ref_spinal_cord) <- ref_spinal_cord$subtype_annotation
```

```{r}
head(ref_spinal_cord@meta.data)

# Motoneurons	- 1331 only
sc_meta %>% 
  group_by(top_level_annotation) %>% 
  dplyr::count() %>% 
  arrange(desc(n)) %>% 
  write.csv(., "yadav_celltypes.csv", row.names = FALSE)

table(sc_meta$top_level_annotation)


table(sc_meta$Visium_ID)
table(sc_meta$sample)
table(ref_spinal_cord$sample, ref_spinal_cord$sex)
```

```{r fig.height=10, fig.width=10}
# pdf("figures/cell_refs/yadav_GSE190442_ref_spinal_cord_subtype_annotation_UMAP.pdf", height = 20, width = 15)
DimPlot(ref_spinal_cord,
        repel = TRUE,
        label.size = 4,
        label=TRUE) +
  theme(legend.position = 'bottom') +
  scale_color_manual(values=colors_vector)
# dev.off()
```

```{r fig.height=10, fig.width=10}
# pdf("figures/cell_refs/yadav_GSE190442_ref_spinal_cord_top_level_annotation_UMAP.pdf", height = 20, width = 15)
DimPlot(ref_spinal_cord,
        group.by = "top_level_annotation",
        repel = TRUE,
        label.size = 4,
        label=TRUE) +
  theme(legend.position = 'bottom') +
  scale_color_manual(values=colors_vector)
# dev.off()
```

```{r fig.height=10, fig.width=10}
# pdf("figures/cell_refs/yadav_GSE190442_ref_spinal_cord_sample_UMAP.pdf", height = 20, width = 15)
DimPlot(ref_spinal_cord,
        group.by = "sample",
        repel = TRUE,
        label.size = 4,
        label=FALSE) +
  theme(legend.position = 'bottom') +
  scale_color_manual(values=colors_vector)
# dev.off()
```

```{r fig.height=10, fig.width=10}
# pdf("figures/cell_refs/yadav_GSE190442_ref_spinal_cord_sex_UMAP.pdf", height = 20, width = 15)
DimPlot(ref_spinal_cord,
        group.by = "sex",
        repel = TRUE,
        label.size = 4,
        label=FALSE) +
  theme(legend.position = 'bottom') +
  scale_color_manual(values=colors_vector)
# dev.off()
```

## Rayon:

* 10x Genomics Single Cell 3′ Chip V3.1 
* Cell Ranger (3.1.0) and the GRCh38-3.0.0

```{r}
embryonic_sc_meta <- read.csv("references/human_single_cell/metadata/human_metadata.csv") %>% 
  mutate(
        barcode =  X,
        dataset_id = str_split_fixed(X, "_",n=2)[,2]) %>% 
  janitor::clean_names() %>% 
  mutate_at(vars(type_step1, type_step2), ~gsub("\\s{1,}","_", .)) %>% 
  mutate_at(vars(type_step2), ~ifelse(is.na(.), type_step1, .)) %>% 
  column_to_rownames("x")

head(embryonic_sc_meta)
# table(embryonic_sc_meta$orig_timepoint)
```

```{r}
embryonic_sc_meta %>% 
  select(type_step1, type_step2)  %>% 
  distinct() %>% 
  View()
```

```{r}
dataset_suffix <- embryonic_sc_meta %>% 
  group_by(orig.ident, dataset_id) %>% 
  dplyr::count() %>% 
  mutate(dataset_order = gsub("CS12_brachial","CS12_spinalcord", orig.ident) %>% 
           as.factor())

# head(dataset_suffix)
```

```{r}
files <- dir("resources/cell_refs/Rayon_2021/GSE171892_RAW", pattern = "^human_.+", full.names = TRUE) 
# head(files)
```

```{r eval=FALSE}
embryonic_sc_counts <- purrr::map(files, function(x){
  print(x)

  obj <- CreateSeuratObject(counts = Read10X(x),
                            project = basename(x))
  obj <- RenameCells(obj, new.names = gsub("-1","",colnames(obj)))

  return(obj)
})
names(embryonic_sc_counts) <- basename(files)
embryonic_sc_counts <- embryonic_sc_counts[dataset_suffix$dataset_order]
# embryonic_sc_counts$human_CS12_spinalcord_rep1@meta.data
# embryonic_sc_counts$human_CS19_thoracic_rep2@meta.data

# saveRDS(embryonic_sc_counts,"resources/cell_refs/Rayon_2021/GSE171892_seurat_obj_list.RDS")

if(!exists("embryonic_sc_counts")){
  embryonic_sc_counts <- readRDS("resources/cell_refs/Rayon_2021/GSE171892_seurat_obj_list.RDS")
}

idx <- grep("human_CS12_spinalcord_rep1",names(embryonic_sc_counts), invert = TRUE)
embryonic_sc <- merge(x = embryonic_sc_counts[["human_CS12_spinalcord_rep1"]], 
                      y = embryonic_sc_counts[idx])
embryonic_sc #33538 features across 120620 samples within 1 assay
embryonic_sc$barcode <- rownames(embryonic_sc@meta.data)

# head(embryonic_sc@meta.data)
# table(embryonic_sc_meta$barcode %in% embryonic_sc$barcode) #TRUE

all_sc_meta <- embryonic_sc@meta.data %>% 
  mutate(dataset_id = str_split_fixed(barcode, "_",n=2)[,2]) %>% 
  group_by(orig.ident, dataset_id) %>%
  dplyr::count()

# all_sc_meta
# dataset_suffix
# head(rownames(embryonic_sc))
# saveRDS(embryonic_sc, "resources/cell_refs/Rayon_2021/GSE171892_merged_seurat_obj_merged.RDS")
```

**Issue with GSM5236521_human_CS14_thoracic_rep1.tar.gz**
--> re-downloading this sample resolved the issue
Using block 1 from human_CS14_thoracic_rep1 to learn model.
Error in make_cell_attr(umi, cell_attr, latent_var, batch_var, latent_var_nonreg,  : 
  cell attribute "log_umi" contains NA, NaN, or infinite value
  
--> this sample has almost zero read counts
53,967,784 Apr 12  2021 GSM5236521_human_CS14_thoracic_rep1.tar.gz
53,967,784
File	GSM5236521_human_CS14_thoracic_rep1.tar.gz	04/12/2021 09:33:42	53,967,784	TAR
  
```{r eval=FALSE}
cell_counts <- purrr::map(Layers(embryonic_sc[["RNA"]]),function(x){
  
  cell_totals <- colSums(embryonic_sc[["RNA"]][[x]])
  cell_totals <- cell_totals[order(cell_totals)]
  
})

no_counts <- purrr::map(cell_counts, function(x) which(x == 0))
names(no_counts) <- Layers(embryonic_sc[["RNA"]])


# length(no_counts$counts.human_CS14_thoracic_rep1)
```

```{r eval=FALSE, fig.height=10, fig.width=10}
embryonic_sc <- subset(embryonic_sc, cells = embryonic_sc_meta$barcode)
embryonic_sc <- AddMetaData(embryonic_sc, metadata = embryonic_sc_meta)
# embryonic_sc #33538 features across 43485 samples within 1 assay

embryonic_sc <- embryonic_sc  %>%
  PercentageFeatureSet(
                       assay="RNA",
                       pattern = "^MT-",
                       col.name = "percent.mt") %>%
  PercentageFeatureSet(
                       assay="RNA",
                       pattern = "^RP[SL]",
                       col.name = "percent.ribo")

VlnPlot(embryonic_sc, 
        pt.size = 0,
        features = c("nCount_RNA","nFeature_RNA","percent.mt", "percent.ribo"),
        ncol = 2)

embryonic_sc <- embryonic_sc %>% 
  SCTransform(
              vst.flavor = "v2",
              method = "glmGamPoi",
              verbose = T) %>% 
    RunPCA(npcs = 50, verbose = T)

embryonic_sc <- embryonic_sc %>% 
    RunUMAP(reduction = "pca",
            dims = 1:50,
            n.components = 3,
            min.dist = 0.3,
            verbose = T) %>%
    FindNeighbors(reduction = "pca",
                  dims = 1:50, verbose = T) %>%
    FindClusters(resolution = 0.8, verbose = T)

embryonic_sc$type_step2 <- ifelse(embryonic_sc$type_step2=="", "unknown", embryonic_sc$type_step2)
# saveRDS(embryonic_sc, "resources/cell_refs/Rayon_2021/GSE171892_merged_postqc_normalized_seurat_obj.RDS")
```

```{r}
embryonic_sc_meta %>% 
  group_by(type_step1) %>% 
  dplyr::count() %>% 
  arrange(desc(n)) %>% 
  # write.csv(.,"resources/cell_refs/Rayon_2021/GSE171892_merged_postqc_normalized_seurat_cell_type_table.csv", row.names = FALSE)
```


```{r}
embryonic_sc@meta.data %>% head()
table(embryonic_sc@meta.data$orig_tissue)
table(embryonic_sc@meta.data$orig.ident)
table(embryonic_sc@meta.data$type_step2)
```

```{r fig.height=10, fig.width=10}
# pdf("figures/cell_refs/rayon_GSE171892_embryonic_spinal_cord_type_step1_UMAP.pdf", height = 20, width = 15)
DimPlot(embryonic_sc,
        group.by = "type_step1",
        repel = TRUE,
        label.size = 4,
        label=TRUE) +
  theme(legend.position = 'bottom') +
  scale_color_manual(values=colors_vector)
# dev.off()
```

```{r fig.height=10, fig.width=10}
# pdf("figures/cell_refs/rayon_GSE171892_embryonic_spinal_cord_type_step2_UMAP.pdf", height = 20, width = 15)
DimPlot(embryonic_sc,
        group.by = "type_step2",
        repel = TRUE,
        label.size = 4,
        label=TRUE) +
  theme(legend.position = 'bottom') +
  scale_color_manual(values=colors_vector)
# dev.off()
```

```{r fig.height=10, fig.width=10}
# pdf("figures/cell_refs/rayon_GSE171892_embryonic_spinal_cord_orig.ident_UMAP.pdf", height = 20, width = 15)
DimPlot(embryonic_sc,
        group.by = "orig.ident",
        repel = TRUE,
        label.size = 4,
        label=TRUE) +
  theme(legend.position = 'bottom') +
  scale_color_manual(values=colors_vector)
# dev.off()
```

```{r fig.height=10, fig.width=10}
# pdf("figures/cell_refs/rayon_GSE171892_embryonic_spinal_cord_sample_UMAP.pdf", height = 20, width = 15)
DimPlot(embryonic_sc,
        group.by = "orig_replicate",
        repel = TRUE,
        label.size = 4,
        label=FALSE) +
  theme(legend.position = 'bottom') +
  scale_color_manual(values=colors_vector)
# dev.off()
```

```{r fig.height=10, fig.width=10}
# pdf("figures/cell_refs/rayon_GSE171892_embryonic_spinal_cord_orig_tissue_UMAP.pdf", height = 20, width = 15)
DimPlot(embryonic_sc,
        group.by = "orig_tissue",
        repel = TRUE,
        label.size = 4,
        label=FALSE) +
  theme(legend.position = 'bottom') +
  scale_color_manual(values=colors_vector)
# dev.off()
```

# Compare Gene Refs

```{r}
head(ref_motor_cortex@assays$RNA)
head(ref_spinal_cord@assays$RNA)

genes_common <- base::intersect(rownames(ref_motor_cortex), rownames(ref_spinal_cord))
length(genes_common) #15,830 only in common :/
```

```{r}
missing_genes <- data.frame(gene_name=rownames(ref_motor_cortex),
           data_set="mc") %>% 
  full_join(., data.frame(gene_name=rownames(ref_spinal_cord),
                          data_set="sc"),
            by="gene_name")

missing_genes %>% 
  filter(is.na(data_set.x))
```

# Integrate

un-normalized counts (layer='counts'), 
normalized data (layer='data'), 
z-scored/variance-stabilized data (layer='scale.data')

https://github.com/satijalab/seurat/issues/7587
if using merge() 
> Error in DelayedArray::colAutoGrid(x = counts, ncol = length(x = layer.cells)) :
> 'ncol' is too big. Blocks of length > .Machine$integer.max are not supported yet. Please specify a smaller 'ncol'.
merged_obj <- merge(x = ref_spinal_cord, 
                    y = ref_motor_cortex,
                    merge.data = FALSE)
                    
```{r}
if(!exists("ref_motor_cortex")){
  ref_motor_cortex  <- readRDS("resources/cell_refs/allen_institute/M1/human_M1_normalized_seurat_object.RDS")
}

if(!exists("ref_spinal_cord")){
  ref_spinal_cord <- readRDS("resources/cell_refs/Yadav_2023/GSE190442_aggregated_postqc_normalized_seurat_obj.RDS")
}

if(!exists("embryonic_sc")){
  embryonic_sc <- readRDS("resources/cell_refs/Rayon_2021/GSE171892_merged_postqc_normalized_seurat_obj.RDS")
}

ref_spinal_cord$donor <- paste0("sample.", ref_spinal_cord$sample)
ref_motor_cortex$donor <- ref_motor_cortex$external_donor_name_label
embryonic_sc$donor <- embryonic_sc$orig_timepoint

# top_level_annotation == subclass_label, type_step1
# subtype_annotation == cluster_label, type_step2
merged_meta <- ref_spinal_cord@meta.data %>% 
  bind_rows(., 
            select(ref_motor_cortex@meta.data, 
                   barcode = sample_name,
                   subtype_annotation = cluster_label ,
                   top_level_annotation = subclass_label, 
                   everything())) %>% 
  bind_rows(., select(embryonic_sc@meta.data,
                      subtype_annotation = type_step2,
                      top_level_annotation = type_step1,
                      everything())) %>% 
  select(barcode, orig.ident:subtype_annotation, donor, everything())

# head(merged_meta)
dim(merged_meta) # 175,307
table(merged_meta$donor)
# write.csv(merged_meta, "results/integrated_ref/Allen_Rayon_Yadav_integrated_cell_ref_metadata.csv", row.names = F)

merged_obj <-  CreateSeuratObject(counts = list("ref_spinal_cord"=LayerData(ref_spinal_cord, layer = 'counts'),
                                                "ref_motor_cortex"=LayerData(ref_motor_cortex, layer = 'counts'),
                                                "ref_embryonic_sc"=LayerData(embryonic_sc, layer = 'counts')), 
                                  meta.data = merged_meta)

merged_obj <- JoinLayers(merged_obj)

merged_obj
rm(ref_spinal_cord, ref_motor_cortex, embryonic_sc)
# table(merged_obj$donor)
# saveRDS(merged_obj, "resources/cell_refs/integrated/embryonic_adult_spinal_cord_motor_cortex_merged_seurat_obj.RDS")
```


SCT integration: 
https://github.com/satijalab/seurat/issues/7542

 https://github.com/satijalab/seurat/issues/7145 - You don't need to worry about those warnings. We will fix them in Seurat-object package.
 Warning: Different cells and/or features from existing assay SCT

--> Also, did not use plan(multisession) though unsure if that was actually causing the errors.

it looks like they string split the original IDs and did NOT provide the actual donor IDs
> table(embryonic_sc_meta$orig_ident, embryonic_sc_meta$orig_replicate)


```{r eval=FALSE}
if(!exists("merged_obj")){
  merged_obj <- read.csv("resources/cell_refs/integrated/embryonic_adult_spinal_cord_motor_cortex_merged_seurat_obj.RDS")
}

merged_split_obj <-  split(merged_obj, f = merged_obj$donor) 
rm(merged_obj)
gc()
gcinfo(TRUE)

merged_split_obj <- merged_split_obj %>% 
  SCTransform(method = "glmGamPoi",
              vst.flavor = "v2", 
              verbose = TRUE) %>% 
  RunPCA(npcs = 50,
         verbose = TRUE)

saveRDS(merged_split_obj,"results/integrated_ref/embryonic_adult_spinal_cord_motor_cortex_merged_split_normalized_seurat_obj.RDS")
# merged_split_obj <- readRDS("results/integrated_ref/embryonic_adult_spinal_cord_motor_cortex_merged_split_normalized_seurat_obj.RDS")

# Add Meta Data
merged_split_obj$dataset <- gsub(".+_CS[0-9]{2}_.+","Rayon_2021",merged_split_obj$orig.ident)
merged_split_obj$top_level_annotation <- gsub("\\s{1,}","_",merged_split_obj$top_level_annotation)
merged_split_obj$subtype_annotation <- gsub("\\s{1,}","_",merged_split_obj$subtype_annotation)
merged_split_obj$motorneuron <- ifelse(grepl("Moto|MN|pMN",merged_split_obj$subtype_annotation), merged_split_obj$subtype_annotation, "Other")
#CS12 and CS17 indicating these samples were male, others female
sc_meta_to_add <- merged_split_obj@meta.data %>% 
  select(-barcode) %>% 
  rownames_to_column("barcode") %>% 
  mutate(sample = as.character(sample)) %>% 
  left_join(., sc_pt_meta, by=c("sample"="donor")) %>% 

  # https://www.ehd.org/virtual-human-embryo/ages.php?stage=1
  mutate(age_clean = case_when(
    dataset == "yadav_2023" ~ paste(age, "yrs", sep="_"),
    grepl("CS12", orig.ident) ~ "29-31_days",
    grepl("CS14", orig.ident) ~  "33-35_days",
    grepl("CS17", orig.ident) ~ "39-42_days",
    grepl("CS19", orig.ident) ~ "45-47_days",
    TRUE ~ "unknown"), 
    sex_clean = case_when(
      dataset == "yadav_2023" ~ sex, 
      dataset == "Allen_M1" ~ donor_sex_label,
      grepl("CS12|CS17", orig.ident) ~ "M",
      grepl("CS14|CS19", orig.ident) ~ "F",
      TRUE ~ "unknown"
    )) %>% 
  select(barcode, age_clean, sex_clean, tissue, cause_of_death) %>% 
  set_rownames(.$barcode)
merged_split_obj <- AddMetaData(merged_split_obj, metadata = sc_meta_to_add)

# Integration
# k.anchor = 5 is default used here
merged_split_obj <- IntegrateLayers(merged_split_obj, 
                              method = RPCAIntegration,
                              orig.reduction = "pca",
                              new.reduction = "integrated.rpca", 
                              normalization.method = 'SCT', #why is this required?
                              assay = "SCT",
                              verbose = TRUE)

merged_split_obj <- IntegrateLayers(merged_split_obj, 
                              method = HarmonyIntegration,
                              orig.reduction = "pca",
                              new.reduction = "harmony", 
                              assay = "SCT",
                              verbose = TRUE)

# Run UMAP
merged_split_obj <- RunUMAP(merged_split_obj,
               reduction = "pca",
               dims = 1:50,
               reduction.name = "umap.unintegrated") %>% 
  RunUMAP(
               reduction = "harmony",
               dims = 1:50,
               reduction.name = "umap.harmony") %>% 
  RunUMAP(
              reduction = "integrated.rpca",
               dims = 1:50,
               reduction.name = "umap.rpca")

# set idents
Idents(merged_split_obj) <- merged_split_obj$orig.ident 

# save object
# saveRDS(merged_split_obj,"results/integrated_ref/embryonic_adult_spinal_cord_motor_cortex_integrated_normalized_seurat_obj.RDS")
```


https://satijalab.org/seurat/articles/seurat5_integration.html

* You will need to do  rejoin the layers before performing any differential expression analysis.
* However, you can always resplit the layers in case you would like to reperform integrative analysis.

```{r}
merged_split_obj <- readRDS("results/integrated_ref/embryonic_adult_spinal_cord_motor_cortex_integrated_normalized_seurat_obj.RDS")


tissue <- merged_split_obj@meta.data %>% 
  mutate(tissue = case_when(
  dataset == "Allen_M1" ~ "adult_motor_cortex",
  dataset == "Rayon_2021" ~  paste0(gsub("human_|_rep[0-9]|spinalcord","", orig.ident), "_spinal_cord"),
  dataset == "yadav_2023" ~ paste0("adult_",tissue, "_spinal_cord")
)) %>% 
  pull(tissue)

merged_split_obj$tissue <- tissue
Idents(merged_split_obj) <- merged_split_obj$dataset
```

```{r}
DefaultAssay(merged_split_obj)
merged_split_obj
```

```{r}
# merged_split_obj$dataset %>% table()
# merged_split_obj$donor %>% table()
# merged_split_obj$motorneuron %>% table() %>% write.csv("mn_table.csv", row.names = F)

merged_split_obj@meta.data %>% 
  group_by(top_level_annotation, subtype_annotation) %>% 
  dplyr::count() %>% 
  arrange(desc(n))

merged_split_obj@meta.data %>% 
  group_by(top_level_annotation) %>% 
  dplyr::count() %>% 
  arrange(desc(n))
```

```{r}
top_level_table_merged <- merged_split_obj@meta.data %>% 
  group_by(dataset,class_label, top_level_annotation, subtype_annotation) %>% 
  dplyr::count() %>% 
  arrange(top_level_annotation)

top_level_table_merged
# write.csv(top_level_table_merged, "results/integrated_ref/Allen_Rayon_Yadav_integrated_celltype_labels_table.csv", row.names = FALSE)
```

```{r}
merged_split_obj@meta.data$
```

```{r}
top_level_table_merged
```


## Figures

```{r fig.height=10, fig.width=11}
# pdf("figures/cell_refs/allen_yadav_radon_integrated/embryonic_adult_spinal_cord_motor_cortex_unintegrated_merged_bySource_UMAP.pdf", height = 10, width = 11)
DimPlot(merged_split_obj, reduction = 'umap.unintegrated') 
# dev.off()

# pdf("figures/cell_refs/allen_yadav_radon_integrated/embryonic_adult_spinal_cord_motor_cortex_integrated_harmony_bySource_UMAP.pdf", height = 10, width = 11)
DimPlot(merged_split_obj, reduction = 'umap.harmony')
# dev.off()

# pdf("figures/cell_refs/allen_yadav_radon_integrated/embryonic_adult_spinal_cord_motor_cortex_integrated_rpca_bySource_UMAP.pdf", height = 10, width = 11)
DimPlot(merged_split_obj, reduction = 'umap.rpca')
# dev.off()
```

```{r fig.height=10, fig.width=11}
colors_mn <- c("Other"="grey80", "Motoneurons" = "#641A80FF", pMN = "#B63679FF" ,MN = "#FECE91FF")
# pdf("figures/cell_refs/embryonic_adult_spinal_cord_motor_cortex_unintegrated_merged_motorneuron_UMAP.pdf", height = 10, width = 11)
DimPlot(merged_split_obj,
        group.by = 'motorneuron',
        reduction = 'umap.unintegrated') +
  scale_color_manual(values = colors_mn)
  # scale_color_viridis_d(option="B", begin = 0.3, end = 0.95)
# dev.off()

# pdf("figures/cell_refs/embryonic_adult_spinal_cord_motor_cortex_integrated_harmony_motorneuron_UMAP.pdf", height = 10, width = 11)
DimPlot(merged_split_obj, 
        group.by = 'motorneuron',
        reduction = 'umap.harmony') +
  scale_color_manual(values = colors_mn)
  # scale_color_viridis_d(option="B", begin = 0.3, end = 0.95)
# dev.off()

# pdf("figures/cell_refs/embryonic_adult_spinal_cord_motor_cortex_integrated_rpca_motorneuron_UMAP.pdf", height = 10, width = 11)
DimPlot(merged_split_obj,
        group.by = 'motorneuron',
        reduction = 'umap.rpca') +
  scale_color_manual(values = colors_mn)
  # scale_color_viridis_d(option="B", begin = 0.3, end = 0.95)
# dev.off()


# pdf("figures/cell_refs/embryonic_adult_spinal_cord_motor_cortex_integrated_rpca_motorneuron_byDataset_UMAP.pdf", height = 7, width = 11)
DimPlot(merged_split_obj,
        group.by = 'motorneuron',
        split.by = "dataset",
        reduction = 'umap.harmony') +
  scale_color_manual(values = colors_mn)
  # scale_color_viridis_d(option="B", begin = 0.3, end = 0.95)
# dev.off()
```

```{r fig.height=10, fig.width=11}
# pdf("figures/cell_refs/embryonic_adult_spinal_cord_motor_cortex_unintegrated_merged_age_clean_UMAP.pdf", height = 10, width = 11)
DimPlot(merged_split_obj,
        group.by = 'age_clean',
        reduction = 'umap.unintegrated') +
  scale_color_manual(values = colors_vector)
# dev.off()

# pdf("figures/cell_refs/embryonic_adult_spinal_cord_motor_cortex_integrated_harmony_age_clean_UMAP.pdf", height = 10, width = 11)
DimPlot(merged_split_obj, 
        group.by = 'age_clean',
        reduction = 'umap.harmony') +
  scale_color_brewer(palette = "Paired")
# dev.off()

# pdf("figures/cell_refs/embryonic_adult_spinal_cord_motor_cortex_integrated_rpca_age_clean_UMAP.pdf", height = 10, width = 11)
DimPlot(merged_split_obj,
        group.by = 'age_clean',
        reduction = 'umap.rpca') +
  scale_color_brewer(palette = "Paired")
# dev.off()
```

```{r fig.height=10, fig.width=11}
# pdf("figures/cell_refs/embryonic_adult_spinal_cord_motor_cortex_unintegrated_merged_byDonor_UMAP.pdf", height = 10, width = 11)
DimPlot(merged_split_obj, 
        group.by = 'donor',
        reduction = 'umap.unintegrated') +
  scale_color_manual(values = colors_vector) 
# dev.off()

# pdf("figures/cell_refs/embryonic_adult_spinal_cord_motor_cortex_integrated_harmony_byDonor_UMAP.pdf", height = 10, width = 11)
DimPlot(merged_split_obj, 
        group.by = 'donor',
        reduction = 'umap.harmony') +
  scale_color_manual(values = colors_vector) 
# dev.off()

# pdf("figures/cell_refs/embryonic_adult_spinal_cord_motor_cortex_integrated_rpca_byDonor_UMAP.pdf", height = 10, width = 11)
DimPlot(merged_split_obj, 
        group.by = 'donor',
        reduction = 'umap.rpca') +
  scale_color_manual(values = colors_vector) 
# dev.off()
```

```{r fig.height=10, fig.width=11}
# pdf("figures/cell_refs/embryonic_adult_spinal_cord_motor_cortex_unintegrated_merged_tissue_UMAP.pdf", height = 15, width = 11)
DimPlot(merged_split_obj, 
        label = F,
        repel = TRUE,
        # label.size = 2.5,
        group.by = 'tissue',
        reduction = 'umap.unintegrated')  +
  scale_color_manual(values = colors_vector) +
  theme(legend.position = 'bottom')
# dev.off()

# pdf("figures/cell_refs/embryonic_adult_spinal_cord_motor_cortex_integrated_harmony_tissue_UMAP.pdf", height = 15, width = 11)
DimPlot(merged_split_obj, 
        label = F,
        repel = TRUE,
        # label.size = 2.5,
        group.by = 'tissue',
        reduction = 'umap.harmony') +
  scale_color_manual(values = colors_vector) +
  theme(legend.position = 'bottom')
# dev.off()

# pdf("figures/cell_refs/embryonic_adult_spinal_cord_motor_cortex_integrated_rpca_tissue_UMAP.pdf", height = 15, width = 11)
DimPlot(merged_split_obj,
        label = F,
        repel = TRUE,
        # label.size = 2.5,
        group.by = 'tissue',
        reduction = 'umap.rpca') +
  scale_color_manual(values = colors_vector) +
  theme(legend.position = 'bottom')
# dev.off()
```

```{r  fig.height=15, fig.width=15}
# pdf("figures/cell_refs/embryonic_adult_spinal_cord_motor_cortex_unintegrated_merged_top_level_annotation_UMAP.pdf", height = 15, width = 11)
DimPlot(merged_split_obj, 
        label = TRUE,
        repel = TRUE,
        # label.size = 2.5,
        group.by = 'top_level_annotation',
        reduction = 'umap.unintegrated')  +
  scale_color_manual(values = colors_vector) +
  theme(legend.position = 'bottom')
# dev.off()

# pdf("figures/cell_refs/embryonic_adult_spinal_cord_motor_cortex_integrated_harmony_top_level_annotation_UMAP.pdf", height = 15, width = 11)
DimPlot(merged_split_obj, 
        label = TRUE,
        repel = TRUE,
        # label.size = 2.5,
        group.by = 'top_level_annotation',
        reduction = 'umap.harmony') +
  scale_color_manual(values = colors_vector) +
  theme(legend.position = 'bottom')
# dev.off()

# pdf("figures/cell_refs/embryonic_adult_spinal_cord_motor_cortex_integrated_rpca_top_level_annotation_UMAP.pdf", height = 15, width = 11)
DimPlot(merged_split_obj,
        label = TRUE,
        repel = TRUE,
        # label.size = 2.5,
        group.by = 'top_level_annotation',
        reduction = 'umap.rpca') +
  scale_color_manual(values = colors_vector) +
  theme(legend.position = 'bottom')
# dev.off()
```

## Motor Neurons and subsets

```{r}
types_across_ds <- unique(c(grep("OPC|Astr|Oligo|Endo", top_level_table_merged$top_level_annotation, value=TRUE, ignore.case = TRUE),
                          grep("MN|pMN|Moto", top_level_table_merged$subtype_annotation, value=TRUE, ignore.case = TRUE)))

cellids <- merged_split_obj@meta.data %>% 
  filter(top_level_annotation %in% types_across_ds | subtype_annotation %in% types_across_ds) %>% 
  pull(barcode)

subset_common_types <- subset(merged_split_obj, cells = cellids)
subset_common_types$celltypes <- subset_common_types@meta.data %>% 
  mutate(celltypes = case_when(
    grepl("MN|pMN|Moto", subtype_annotation, ignore.case = TRUE) ~ subtype_annotation,
    TRUE ~ top_level_annotation
  )) %>% 
  pull(celltypes)

Idents(subset_common_types) <- subset_common_types$celltypes
Idents(merged_split_obj) <- merged_split_obj$top_level_annotation
```

```{r fig.height=10, fig.width=10}
# pdf("figures/cell_refs/embryonic_adult_spinal_cord_motor_cortex_integrated_harmony_subset_UMAP.pdf", height = 15, width = 11)
DimPlot(subset_common_types, 
        label = TRUE,
        repel = TRUE,
        group.by = 'celltypes',
        # split.by = 'dataset',
        reduction = 'umap.harmony') +
  scale_color_manual(values = colors_vector) +
  theme(legend.position = 'bottom')
# dev.off()

# pdf("figures/cell_refs/embryonic_adult_spinal_cord_motor_cortex_integrated_rpca_subset_UMAP.pdf", height = 15, width = 11)
DimPlot(subset_common_types,
        label = TRUE,
        repel = TRUE,
        group.by = 'celltypes',
        reduction = 'umap.rpca') +
  scale_color_manual(values = colors_vector) +
  theme(legend.position = 'bottom')
# dev.off()
```

## Marker Gene Expression 

```{r}
# sel_markers <- names(rayon_marker_gene_list) %>%  grep("pMN|MN|rayon_Oligodendrocyte_NA", ., value = T)
# sel_markers
```

```{r}
mn_feature_plots <- purrr::map(c("umap.rpca","umap.harmony"), function(red){
  purrr::map(sel_markers, function(x){
      out1 <- glue("figures/cell_refs/allen_yadav_radon_integrated/embryonic_adult_sc_mc_integrated_{red}_{x}_UMAP.png")
      fp1 <- FeaturePlot(merged_split_obj,
                       label = TRUE, 
                       repel = TRUE,
                       label.size = 2,
                       features = rayon_marker_gene_list[[x]],
                       reduction = red) +
      patchwork::plot_annotation(title = x)
    
      out2 <- glue("figures/cell_refs/allen_yadav_radon_integrated/embryonic_adult_sc_mc_integrated_{red}_{x}_subset_UMAP.png")
      fp2 <- FeaturePlot(subset_common_types,
                features = rayon_marker_gene_list[[x]],
                label = TRUE,
                repel = TRUE,
                reduction = red) +
      patchwork::plot_annotation(title = x)
      
      
      png(out1, units = "in", res = 100, height = 20, width = 20)
      print(fp1)
      dev.off()
      
      png(out2, units = "in", res = 100, height = 20, width = 20)
      print(fp2)
      dev.off()
  })

})
```








# Session Info

```{r}
session_info()
```

